// ng-zorro-antd-mobile/esm5/white-space/white-space.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from '@angular/core';
/** @typedef {?} */
var sizeType;
export { sizeType };
var WhiteSpace = /** @class */ (function () {
    function WhiteSpace() {
        this.prefixCls = 'am-whitespace';
        this.size = 'md';
        this.amWhiteSpace = true;
    }
    Object.defineProperty(WhiteSpace.prototype, "amWhitespaceXs", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'xs';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WhiteSpace.prototype, "amWhitespaceSm", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'sm';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WhiteSpace.prototype, "amWhitespaceMd", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'md';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WhiteSpace.prototype, "amWhitespaceLg", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'lg';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WhiteSpace.prototype, "amWhitespaceXl", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'xl';
        },
        enumerable: true,
        configurable: true
    });
    WhiteSpace.decorators = [
        { type: Component, args: [{
                    selector: 'WhiteSpace, nzm-whitespace',
                    template: ""
                }] }
    ];
    /** @nocollapse */
    WhiteSpace.ctorParameters = function () { return []; };
    WhiteSpace.propDecorators = {
        size: [{ type: Input }],
        amWhiteSpace: [{ type: HostBinding, args: ['class.am-whitespace',] }],
        amWhitespaceXs: [{ type: HostBinding, args: ['class.am-whitespace-xs',] }],
        amWhitespaceSm: [{ type: HostBinding, args: ['class.am-whitespace-sm',] }],
        amWhitespaceMd: [{ type: HostBinding, args: ['class.am-whitespace-md',] }],
        amWhitespaceLg: [{ type: HostBinding, args: ['class.am-whitespace-lg',] }],
        amWhitespaceXl: [{ type: HostBinding, args: ['class.am-whitespace-xl',] }]
    };
    return WhiteSpace;
}());
export { WhiteSpace };
if (false) {
    /** @type {?} */
    WhiteSpace.prototype.prefixCls;
    /** @type {?} */
    WhiteSpace.prototype.size;
    /** @type {?} */
    WhiteSpace.prototype.amWhiteSpace;
}
