// ng-zorro-antd-mobile/esm5/slider/slider-track/slider-track.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
var SliderTrack = /** @class */ (function () {
    function SliderTrack(_elf, _sanitizer) {
        this._elf = _elf;
        this._sanitizer = _sanitizer;
        this.prefixCls = 'am-slider';
        this._included = true;
    }
    Object.defineProperty(SliderTrack.prototype, "className", {
        get: /**
         * @return {?}
         */
        function () {
            return this._className;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._className = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderTrack.prototype, "included", {
        get: /**
         * @return {?}
         */
        function () {
            return this._included;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._included = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderTrack.prototype, "offset", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._offset = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderTrack.prototype, "length", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._length = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SliderTrack.prototype, "style", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._style = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SliderTrack.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var positonStyle = {
            left: this._offset + "%",
            width: this._length + "%"
        };
        this.elStyle = tslib_1.__assign({}, this._style, positonStyle);
    };
    SliderTrack.decorators = [
        { type: Component, args: [{
                    selector: 'SliderTrack, nzm-slider-track',
                    template: "<div *ngIf=\"included\" \n     [ngClass]=\"className\" \n     [ngStyle]=\"elStyle\" >\n</div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    SliderTrack.ctorParameters = function () { return [
        { type: ElementRef },
        { type: DomSanitizer }
    ]; };
    SliderTrack.propDecorators = {
        className: [{ type: Input }],
        included: [{ type: Input }],
        offset: [{ type: Input }],
        length: [{ type: Input }],
        style: [{ type: Input }]
    };
    return SliderTrack;
}());
export { SliderTrack };
if (false) {
    /** @type {?} */
    SliderTrack.prototype.prefixCls;
    /** @type {?} */
    SliderTrack.prototype.elStyle;
    /** @type {?} */
    SliderTrack.prototype._className;
    /** @type {?} */
    SliderTrack.prototype._included;
    /** @type {?} */
    SliderTrack.prototype._offset;
    /** @type {?} */
    SliderTrack.prototype._length;
    /** @type {?} */
    SliderTrack.prototype._style;
    /** @type {?} */
    SliderTrack.prototype._elf;
    /** @type {?} */
    SliderTrack.prototype._sanitizer;
}
