// ng-zorro-antd-mobile/esm2015/segmented-control/segmented-control.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, HostBinding, EventEmitter } from '@angular/core';
/**
 * @record
 */
export function OnChangeEvent() { }
/** @type {?} */
OnChangeEvent.prototype.selectedIndex;
/** @type {?} */
OnChangeEvent.prototype.value;
export class SegmentedControlComponent {
    constructor() {
        this.prefixCls = 'am-segment';
        this.tintColor = '#2DB7F5';
        this.disabled = false;
        this.selectedIndex = 0;
        this.onChange = new EventEmitter();
        this.role = 'tablist';
        this.amSegment = true;
    }
    /**
     * @return {?}
     */
    get amDisabled() {
        return this.disabled;
    }
    /**
     * @param {?} index
     * @param {?} value
     * @return {?}
     */
    onClick(index, value) {
        if (!this.disabled && index !== this.selectedIndex) {
            this.selectedIndex = index;
            this.onChange.emit({ selectedIndex: index, value: value });
        }
    }
}
SegmentedControlComponent.decorators = [
    { type: Component, args: [{
                selector: 'SegmentedControl, nz-segmented-control',
                template: "<div role=\"tab\"\n     *ngFor=\"let value of values; let i = index;\"\n     class=\"{{prefixCls}}-item\"\n     [ngClass]=\"{'am-segment-item-selected' : i === selectedIndex}\"\n     [ngStyle]=\"{'border-color': tintColor, 'color': i === selectedIndex ? '#fff' : tintColor, 'background-color': i === selectedIndex ? tintColor : 'transparent'}\"\n     (click)=\"onClick(i, value)\"\n>\n  <div class=\"{{prefixCls}}-item-inner\"\n       [ngStyle]=\"{'background-color': i === selectedIndex ? tintColor : 'transparent'}\"\n  ></div>\n  {{value}}\n</div>\n"
            }] }
];
/** @nocollapse */
SegmentedControlComponent.ctorParameters = () => [];
SegmentedControlComponent.propDecorators = {
    tintColor: [{ type: Input }],
    disabled: [{ type: Input }],
    selectedIndex: [{ type: Input }],
    values: [{ type: Input }],
    onChange: [{ type: Output }],
    role: [{ type: HostBinding, args: ['attr.role',] }],
    amSegment: [{ type: HostBinding, args: ['class.am-segment',] }],
    amDisabled: [{ type: HostBinding, args: ['class.am-segment-disabled',] }]
};
if (false) {
    /** @type {?} */
    SegmentedControlComponent.prototype.prefixCls;
    /** @type {?} */
    SegmentedControlComponent.prototype.tintColor;
    /** @type {?} */
    SegmentedControlComponent.prototype.disabled;
    /** @type {?} */
    SegmentedControlComponent.prototype.selectedIndex;
    /** @type {?} */
    SegmentedControlComponent.prototype.values;
    /** @type {?} */
    SegmentedControlComponent.prototype.onChange;
    /** @type {?} */
    SegmentedControlComponent.prototype.role;
    /** @type {?} */
    SegmentedControlComponent.prototype.amSegment;
}
