// ng-zorro-antd-mobile/esm5/modal/modal-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var BaseOptions = /** @class */ (function () {
    function BaseOptions() {
        this.visible = false;
        this.focus = true;
        this.closable = false;
        this.maskClosable = true;
        this.transparent = false;
        this.popup = false;
        this.transitionName = 'am-zoom';
        this.maskTransitionName = 'am-fade';
    }
    return BaseOptions;
}());
export { BaseOptions };
if (false) {
    /** @type {?} */
    BaseOptions.prototype.visible;
    /** @type {?} */
    BaseOptions.prototype.focus;
    /** @type {?} */
    BaseOptions.prototype.closable;
    /** @type {?} */
    BaseOptions.prototype.maskClosable;
    /** @type {?} */
    BaseOptions.prototype.onClose;
    /** @type {?} */
    BaseOptions.prototype.transparent;
    /** @type {?} */
    BaseOptions.prototype.popup;
    /** @type {?} */
    BaseOptions.prototype.animationType;
    /** @type {?} */
    BaseOptions.prototype.title;
    /** @type {?} */
    BaseOptions.prototype.footer;
    /** @type {?} */
    BaseOptions.prototype.platform;
    /** @type {?} */
    BaseOptions.prototype.className;
    /** @type {?} */
    BaseOptions.prototype.wrapClassName;
    /** @type {?} */
    BaseOptions.prototype.message;
    /** @type {?} */
    BaseOptions.prototype.actions;
    /** @type {?} */
    BaseOptions.prototype.callbackOrActions;
    /** @type {?} */
    BaseOptions.prototype.type;
    /** @type {?} */
    BaseOptions.prototype.defaultValue;
    /** @type {?} */
    BaseOptions.prototype.placeholders;
    /** @type {?} */
    BaseOptions.prototype.operation;
    /** @type {?} */
    BaseOptions.prototype.transitionName;
    /** @type {?} */
    BaseOptions.prototype.maskTransitionName;
    /** @type {?} */
    BaseOptions.prototype.close;
}
var ModalOptions = /** @class */ (function (_super) {
    tslib_1.__extends(ModalOptions, _super);
    function ModalOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ModalOptions.decorators = [
        { type: Injectable }
    ];
    return ModalOptions;
}(BaseOptions));
export { ModalOptions };
var AlertOptions = /** @class */ (function (_super) {
    tslib_1.__extends(AlertOptions, _super);
    function AlertOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AlertOptions.decorators = [
        { type: Injectable }
    ];
    return AlertOptions;
}(BaseOptions));
export { AlertOptions };
if (false) {
    /**
     * 	提示信息
     * @type {?}
     */
    AlertOptions.prototype.message;
    /**
     * 按钮组, [{text, onPress, style}]
     * @type {?}
     */
    AlertOptions.prototype.actions;
}
var PromptOptions = /** @class */ (function (_super) {
    tslib_1.__extends(PromptOptions, _super);
    function PromptOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PromptOptions.decorators = [
        { type: Injectable }
    ];
    return PromptOptions;
}(BaseOptions));
export { PromptOptions };
if (false) {
    /**
     * 	提示信息
     * @type {?}
     */
    PromptOptions.prototype.message;
    /**
     * 按钮组 [{text, onPress}]
     * 或回调函数 Array or Function
     * @type {?}
     */
    PromptOptions.prototype.callbackOrActions;
    /**
     * prompt 的样式
     * String (default, secure-text, login-password)
     * @type {?}
     */
    PromptOptions.prototype.type;
    /**
     * 默认值(input 为 password 类型不支持)
     * @type {?}
     */
    PromptOptions.prototype.defaultValue;
    /**
     * ['', '']
     * @type {?}
     */
    PromptOptions.prototype.placeholders;
}
var OperationOptions = /** @class */ (function (_super) {
    tslib_1.__extends(OperationOptions, _super);
    function OperationOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OperationOptions.decorators = [
        { type: Injectable }
    ];
    return OperationOptions;
}(BaseOptions));
export { OperationOptions };
if (false) {
    /**
     * 按钮组, [{text, onPress, style}]
     * @type {?}
     */
    OperationOptions.prototype.actions;
}
var Action = /** @class */ (function () {
    function Action() {
    }
    Action.decorators = [
        { type: Injectable }
    ];
    return Action;
}());
export { Action };
if (false) {
    /** @type {?} */
    Action.prototype.text;
    /** @type {?} */
    Action.prototype.onPress;
    /** @type {?} */
    Action.prototype.style;
}
