// ng-zorro-antd-mobile/esm5/card/card-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding, TemplateRef } from '@angular/core';
var CardFooter = /** @class */ (function () {
    function CardFooter() {
        this.prefixCls = 'am-card-footer';
        this.content = null;
        this.extra = null;
        this.cardFooterWrapper = true;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    CardFooter.prototype.isTemplateRef = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof TemplateRef;
    };
    CardFooter.decorators = [
        { type: Component, args: [{
                    selector: 'CardFooter, nzm-card-footer',
                    template: "<div class=\"{{prefixCls}}-content\">\n  <ng-container *ngIf=\"!isTemplateRef(content);else contentTemplate\">{{content}}</ng-container>\n</div>\n<div *ngIf=\"extra\" class=\"{{prefixCls}}-extra\">\n  <ng-container *ngIf=\"!isTemplateRef(extra);else extraTemplate\">{{extra}}</ng-container>\n</div>\n<ng-template #contentTemplate>\n  <ng-template [ngTemplateOutlet]=\"content\"></ng-template>\n</ng-template>\n<ng-template #extraTemplate>\n  <ng-template [ngTemplateOutlet]=\"extra\"></ng-template>\n</ng-template>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    CardFooter.ctorParameters = function () { return []; };
    CardFooter.propDecorators = {
        content: [{ type: Input }],
        extra: [{ type: Input }],
        cardFooterWrapper: [{ type: HostBinding, args: ['class.am-card-footer',] }]
    };
    return CardFooter;
}());
export { CardFooter };
if (false) {
    /** @type {?} */
    CardFooter.prototype.prefixCls;
    /** @type {?} */
    CardFooter.prototype.content;
    /** @type {?} */
    CardFooter.prototype.extra;
    /** @type {?} */
    CardFooter.prototype.cardFooterWrapper;
}
