// ng-zorro-antd-mobile/esm5/card/card.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
var Card = /** @class */ (function () {
    function Card() {
        this.full = false;
        this.cardWrapper = true;
    }
    Object.defineProperty(Card.prototype, "cardFull", {
        get: /**
         * @return {?}
         */
        function () {
            return this.full;
        },
        enumerable: true,
        configurable: true
    });
    Card.decorators = [
        { type: Component, args: [{
                    selector: 'Card, nzm-card',
                    template: "<ng-content></ng-content>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    Card.ctorParameters = function () { return []; };
    Card.propDecorators = {
        full: [{ type: Input }],
        cardWrapper: [{ type: HostBinding, args: ['class.am-card',] }],
        cardFull: [{ type: HostBinding, args: ['class.am-card-full',] }]
    };
    return Card;
}());
export { Card };
if (false) {
    /** @type {?} */
    Card.prototype.full;
    /** @type {?} */
    Card.prototype.cardWrapper;
}
