// ng-zorro-antd-mobile/esm2015/tab-bar/tab-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ContentChildren, QueryList, HostBinding } from '@angular/core';
import { TabPane } from '../tabs/tab-pane.component';
import { TabBarTab } from './tab-bar-tab.component';
/** @typedef {?} */
var TabBarPositionType;
export { TabBarPositionType };
export class TabBar {
    constructor() {
        this.prefixCls = 'am-tab-bar';
        this._tintColor = '#108ee9';
        this._unselectedTintColor = '#888';
        this.hidden = false;
        this.activeTab = 0;
        this.barTintColor = 'white';
        this.tabBarPosition = 'bottom';
        this.tabBar = true;
    }
    /**
     * @return {?}
     */
    get tintColor() {
        return this._tintColor;
    }
    /**
     * @param {?} color
     * @return {?}
     */
    set tintColor(color) {
        this._tintColor = color;
        if (this.tabBarTabs && this.tabBarTabs.length > 0) {
            this.tabBarTabs.forEach((tabBarTab) => {
                tabBarTab.tintColor = this._tintColor;
            });
        }
    }
    /**
     * @return {?}
     */
    get unselectedTintColor() {
        return this._unselectedTintColor;
    }
    /**
     * @param {?} color
     * @return {?}
     */
    set unselectedTintColor(color) {
        this._unselectedTintColor = color;
        if (this.tabBarTabs && this.tabBarTabs.length > 0) {
            this.tabBarTabs.forEach((tabBarTab) => {
                tabBarTab.unselectedTintColor = this._unselectedTintColor;
            });
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    selectTabPane(index) {
        this.tabPanes.forEach((tabPane, indexKey) => {
            if (index < indexKey) {
                tabPane.position = 'right-without-animation';
            }
            else if (index > indexKey) {
                tabPane.position = 'left-without-animation';
            }
            else {
                tabPane.position = 'center-without-animation';
            }
            if (index !== indexKey) {
                tabPane.active = false;
            }
            else {
                tabPane.active = true;
            }
        });
        if (this.tabBarTabs && this.tabBarTabs.length > 0) {
            this.tabBarTabs.forEach((tabBarTab) => {
                tabBarTab.selected = false;
            });
            this.tabBarTabs.toArray()[index].selected = true;
            this.tabBarTabs.toArray()[index].onPress.emit({
                title: this.tabBarTabs.toArray()[index].title,
                key: this.tabBarTabs.toArray()[index].key
            });
        }
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.selectTabPane(this.activeTab);
        this.tabPanes.forEach((tabPane) => {
            tabPane.tintColor = this.tintColor;
            tabPane.unselectedTintColor = this.unselectedTintColor;
        });
    }
}
TabBar.decorators = [
    { type: Component, args: [{
                selector: 'TabBar, nzm-tab-bar',
                template: "<Tabs [activeTab]=\"activeTab\"\n      [swipeable]=\"false\"\n      [useOnPan]=\"false\"\n      [animated]=\"false\"\n      [tabDirection]=\"'horizontal'\"\n      [tabBarPosition]='tabBarPosition'\n      [renderTabBar]=\"TabBarBar\"\n      [tabPanesComponent]=\"tabPanes\"\n>\n  <ng-content></ng-content>\n</Tabs>\n\n<ng-template #TabBarBar>\n  <div class=\"am-tabs-tab-bar-wrap\">\n    <div class=\"{{prefixCls}}-bar\"\n         [ngClass]=\"{'am-tab-bar-bar-hidden-top': 'top' === tabBarPosition && hidden, 'am-tab-bar-bar-hidden-bottom': 'bottom' === tabBarPosition && hidden}\"\n         [style.background-color]=\"barTintColor\">\n      <div *ngFor=\"let tabPane of tabPanes; let i = index;\" (click)=\"selectTabPane(i)\">\n        <ng-container [ngTemplateOutlet]=\"tabPane.title\"></ng-container>\n      </div>\n    </div>\n  </div>\n</ng-template>"
            }] }
];
/** @nocollapse */
TabBar.ctorParameters = () => [];
TabBar.propDecorators = {
    tabPanes: [{ type: ContentChildren, args: [TabPane, { descendants: true },] }],
    tabBarTabs: [{ type: ContentChildren, args: [TabBarTab, { descendants: true },] }],
    hidden: [{ type: Input }],
    activeTab: [{ type: Input }],
    barTintColor: [{ type: Input }],
    tabBarPosition: [{ type: Input }],
    tintColor: [{ type: Input }],
    unselectedTintColor: [{ type: Input }],
    tabBar: [{ type: HostBinding, args: ['class.am-tab-bar',] }]
};
if (false) {
    /** @type {?} */
    TabBar.prototype.prefixCls;
    /** @type {?} */
    TabBar.prototype._tintColor;
    /** @type {?} */
    TabBar.prototype._unselectedTintColor;
    /** @type {?} */
    TabBar.prototype.tabPanes;
    /** @type {?} */
    TabBar.prototype.tabBarTabs;
    /** @type {?} */
    TabBar.prototype.hidden;
    /** @type {?} */
    TabBar.prototype.activeTab;
    /** @type {?} */
    TabBar.prototype.barTintColor;
    /** @type {?} */
    TabBar.prototype.tabBarPosition;
    /** @type {?} */
    TabBar.prototype.tabBar;
}
