// ng-zorro-antd-mobile/esm5/radio/radio.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, forwardRef, HostBinding, HostListener, EventEmitter, ViewEncapsulation } from '@angular/core';
import { toBoolean } from '../core/util/convert';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * @record
 */
export function OnChangeEvent() { }
/** @type {?} */
OnChangeEvent.prototype.name;
/** @type {?} */
OnChangeEvent.prototype.value;
/** @type {?} */
OnChangeEvent.prototype.checked;
var Radio = /** @class */ (function () {
    function Radio() {
        var _a;
        this.prefixCls = 'am-radio';
        this.classMap = (_a = {},
            _a[this.prefixCls] = true,
            _a[this.prefixCls + "-checked"] = this.checked,
            _a[this.prefixCls + "-disabled"] = this.disabled,
            _a);
        this._checked = false;
        this._disabled = false;
        this._ngModelOnChange = Function.prototype;
        this._ngModelOnTouched = Function.prototype;
        this.onChange = new EventEmitter();
        this.radioWrapper = true;
    }
    Object.defineProperty(Radio.prototype, "checked", {
        get: /**
         * @return {?}
         */
        function () {
            return this._checked;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._checked = value;
            this.updateClassMap();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Radio.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = toBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    Radio.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        if (!this._disabled && !this._checked) {
            this.updateValue(true);
        }
    };
    /**
     * @param {?} checkValue
     * @return {?}
     */
    Radio.prototype.updateValue = /**
     * @param {?} checkValue
     * @return {?}
     */
    function (checkValue) {
        this._ngModelOnChange(checkValue);
        this.onChange.emit({
            name: this.name,
            value: this.value,
            checked: checkValue
        });
        this.checked = checkValue;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    Radio.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (null !== value) {
            this.checked = value;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    Radio.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    Radio.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnTouched = fn;
    };
    /**
     * @return {?}
     */
    Radio.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.updateClassMap();
    };
    /**
     * @return {?}
     */
    Radio.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.updateClassMap();
    };
    /**
     * @return {?}
     */
    Radio.prototype.updateClassMap = /**
     * @return {?}
     */
    function () {
        var _a;
        this.classMap = (_a = {},
            _a[this.prefixCls] = true,
            _a[this.prefixCls + "-checked"] = this.checked,
            _a[this.prefixCls + "-disabled"] = this.disabled,
            _a);
    };
    Radio.decorators = [
        { type: Component, args: [{
                    selector: '[Radio], [nzm-radio]',
                    template: "<span [ngClass]=\"classMap\">\n  <input type=\"radio\"\n         class=\"{{prefixCls}}-input\"\n         [attr.name]=\"name\"\n         [attr.value]=\"value\"\n         [checked]=\"checked\"\n  >\n  <span class=\"{{prefixCls}}-inner\"></span>\n</span>\n<ng-content></ng-content>\n",
                    preserveWhitespaces: false,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return Radio; }),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    Radio.ctorParameters = function () { return []; };
    Radio.propDecorators = {
        name: [{ type: Input }],
        value: [{ type: Input }],
        checked: [{ type: Input }],
        disabled: [{ type: Input }],
        onChange: [{ type: Output }],
        radioWrapper: [{ type: HostBinding, args: ['class.am-radio-wrapper',] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return Radio;
}());
export { Radio };
if (false) {
    /** @type {?} */
    Radio.prototype.prefixCls;
    /** @type {?} */
    Radio.prototype.classMap;
    /** @type {?} */
    Radio.prototype._checked;
    /** @type {?} */
    Radio.prototype._disabled;
    /** @type {?} */
    Radio.prototype._ngModelOnChange;
    /** @type {?} */
    Radio.prototype._ngModelOnTouched;
    /** @type {?} */
    Radio.prototype.name;
    /** @type {?} */
    Radio.prototype.value;
    /** @type {?} */
    Radio.prototype.onChange;
    /** @type {?} */
    Radio.prototype.radioWrapper;
}
