// ng-zorro-antd-mobile/esm5/slider/slider.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef, ViewEncapsulation, HostBinding } from '@angular/core';
var Slider = /** @class */ (function () {
    function Slider(_elf) {
        this._elf = _elf;
        this.prefixCls = 'am-slider';
        this.offset = 0;
        this.length = 0;
        this._min = 0;
        this._max = 100;
        this._step = 1;
        this._value = 10;
        this._disabled = false;
        this._marks = {};
        this._dots = false;
        this._included = true;
        this._trackStyle = {};
        this.onChange = new EventEmitter();
        this.onAfterChange = new EventEmitter();
        this.amSliderWrapper = true;
    }
    Object.defineProperty(Slider.prototype, "min", {
        get: /**
         * @return {?}
         */
        function () {
            return this._min;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._min = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "max", {
        get: /**
         * @return {?}
         */
        function () {
            return this._max;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._max = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "step", {
        get: /**
         * @return {?}
         */
        function () {
            return this._step;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._step = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._value = value;
            this.valueRange();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "defaultValue", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._defaultValue = value;
            this._value = this._defaultValue;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "marks", {
        get: /**
         * @return {?}
         */
        function () {
            return this._marks;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._marks = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "dots", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dots;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._dots = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "included", {
        get: /**
         * @return {?}
         */
        function () {
            return this._included;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._included = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "handleStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._handleStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._handleStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "trackStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._trackStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._trackStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "railStyle", {
        get: /**
         * @return {?}
         */
        function () {
            return this._railStyle;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._railStyle = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Slider.prototype.setCls = /**
     * @return {?}
     */
    function () {
        var _a;
        this.sliderCls = (_a = {},
            _a[this.prefixCls + "-disabled"] = this._disabled,
            _a);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    Slider.prototype.handleChange = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        setTimeout(function () {
            _this.offset = 0;
            _this.length = ((e - _this._min) * 100) / (_this._max - _this._min);
            _this._value = e;
        }, 10);
        this.onChange.emit(e);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    Slider.prototype.handleAfterChange = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var _this = this;
        setTimeout(function () {
            _this.offset = 0;
            _this.length = ((e - _this._min) * 100) / (_this._max - _this._min);
            _this._value = e;
        }, 10);
        this.onAfterChange.emit(e);
    };
    /**
     * @return {?}
     */
    Slider.prototype.valueRange = /**
     * @return {?}
     */
    function () {
        if (this._value < this._min) {
            this._value = this._min;
        }
        if (this._value > this._max) {
            this._value = this._max;
        }
    };
    /**
     * @return {?}
     */
    Slider.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setCls();
        this.valueRange();
        /** @type {?} */
        var sliderCoords = this._elf.nativeElement.getElementsByClassName('am-slider')[0].getBoundingClientRect();
        this.sliderLength = sliderCoords.width;
        this.sliderStart = sliderCoords.left;
    };
    Slider.decorators = [
        { type: Component, args: [{
                    selector: 'Slider , nzm-slider',
                    template: "<div class=\"am-slider\" [ngClass]=\"sliderCls\">\n  <div class=\"am-slider-rail\" [ngStyle]=\"railStyle\"></div>\n  <SliderTrack [className]=\"'am-slider-track'\"\n               [style]=\"trackStyle\"\n               [offset]=\"offset\"\n               [length]=\"length\"\n               [included]=\"included\"\n  ></SliderTrack>\n  <SliderSteps [max]=\"max\"\n               [min]=\"min\"\n               [dots]=\"dots\"\n               [step]=\"step\"\n               [marks]=\"marks\"\n               [lowerBound]=\"min\"\n               [upperBound]=\"value\"\n               [included]=\"included\"\n  ></SliderSteps>\n  <SliderHandle [max]=\"max\"\n                [min]=\"min\"\n                [value]=\"value\"\n                [step]=\"step\"\n                [disabled]=\"disabled\"\n                [handleStyle]=\"handleStyle\"\n                [sliderStart]=\"sliderStart\"\n                [sliderLength]=\"sliderLength\"\n                (onChange)=\"handleChange($event)\"\n                (onAfterChange)=\"handleAfterChange($event)\"\n  ></SliderHandle>\n  <SliderMarks [max]=\"max\"\n               [min]=\"min\"\n               [marks]=\"marks\"\n               [lowerBound]=\"min\"\n               [upperBound]=\"value\"\n               [included]=\"included\"\n  ></SliderMarks>\n</div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    Slider.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    Slider.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        step: [{ type: Input }],
        value: [{ type: Input }],
        defaultValue: [{ type: Input }],
        disabled: [{ type: Input }],
        marks: [{ type: Input }],
        dots: [{ type: Input }],
        included: [{ type: Input }],
        handleStyle: [{ type: Input }],
        trackStyle: [{ type: Input }],
        railStyle: [{ type: Input }],
        onChange: [{ type: Output }],
        onAfterChange: [{ type: Output }],
        amSliderWrapper: [{ type: HostBinding, args: ['class.am-slider-wrapper',] }]
    };
    return Slider;
}());
export { Slider };
if (false) {
    /** @type {?} */
    Slider.prototype.prefixCls;
    /** @type {?} */
    Slider.prototype.sliderLength;
    /** @type {?} */
    Slider.prototype.sliderStart;
    /** @type {?} */
    Slider.prototype.sliderCls;
    /** @type {?} */
    Slider.prototype.offset;
    /** @type {?} */
    Slider.prototype.length;
    /** @type {?} */
    Slider.prototype._min;
    /** @type {?} */
    Slider.prototype._max;
    /** @type {?} */
    Slider.prototype._step;
    /** @type {?} */
    Slider.prototype._value;
    /** @type {?} */
    Slider.prototype._defaultValue;
    /** @type {?} */
    Slider.prototype._disabled;
    /** @type {?} */
    Slider.prototype._marks;
    /** @type {?} */
    Slider.prototype._dots;
    /** @type {?} */
    Slider.prototype._included;
    /** @type {?} */
    Slider.prototype._handleStyle;
    /** @type {?} */
    Slider.prototype._trackStyle;
    /** @type {?} */
    Slider.prototype._railStyle;
    /** @type {?} */
    Slider.prototype.onChange;
    /** @type {?} */
    Slider.prototype.onAfterChange;
    /** @type {?} */
    Slider.prototype.amSliderWrapper;
    /** @type {?} */
    Slider.prototype._elf;
}
