// ng-zorro-antd-mobile/esm5/progress/progress.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
/** @typedef {?} */
var PositionType;
export { PositionType };
var ProgressComponent = /** @class */ (function () {
    function ProgressComponent() {
        this.prefixCls = 'am-progress';
        this._percent = 0;
        this._exceedance = false;
        this.unfilled = true;
        this.position = 'fixed';
        this.barStyle = {};
        this.max = 100;
        this.outer = true;
    }
    Object.defineProperty(ProgressComponent.prototype, "percent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._percent;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._percent = value;
            if (value > 100) {
                this._exceedance = true;
            }
            else {
                this._exceedance = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.percent;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressComponent.prototype, "fixOuter", {
        get: /**
         * @return {?}
         */
        function () {
            return 'fixed' === this.position;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressComponent.prototype, "hideOuter", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.unfilled && !this._exceedance;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressComponent.prototype, "exceedance", {
        get: /**
         * @return {?}
         */
        function () {
            return this._exceedance;
        },
        enumerable: true,
        configurable: true
    });
    ProgressComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Progress, nzm-progress',
                    template: "<div class=\"{{prefixCls}}-bar\" [ngStyle]=\"barStyle\" [style.width.%]=\"percent\"></div>",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    ProgressComponent.ctorParameters = function () { return []; };
    ProgressComponent.propDecorators = {
        unfilled: [{ type: Input }],
        position: [{ type: Input }],
        barStyle: [{ type: Input }],
        percent: [{ type: Input }],
        max: [{ type: HostBinding, args: ['attr.max',] }],
        value: [{ type: HostBinding, args: ['attr.value',] }],
        outer: [{ type: HostBinding, args: ['class.am-progress-outer',] }],
        fixOuter: [{ type: HostBinding, args: ['class.am-progress-fixed-outer',] }],
        hideOuter: [{ type: HostBinding, args: ['class.am-progress-hide-outer',] }],
        exceedance: [{ type: HostBinding, args: ['class.am-progress-exceedance',] }]
    };
    return ProgressComponent;
}());
export { ProgressComponent };
if (false) {
    /** @type {?} */
    ProgressComponent.prototype.prefixCls;
    /** @type {?} */
    ProgressComponent.prototype._percent;
    /** @type {?} */
    ProgressComponent.prototype._exceedance;
    /** @type {?} */
    ProgressComponent.prototype.unfilled;
    /** @type {?} */
    ProgressComponent.prototype.position;
    /** @type {?} */
    ProgressComponent.prototype.barStyle;
    /** @type {?} */
    ProgressComponent.prototype.max;
    /** @type {?} */
    ProgressComponent.prototype.outer;
}
