// ng-zorro-antd-mobile/esm2015/picker/picker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { PickerComponent } from './picker.component';
import { PickerOptions } from './picker-options.provider';
import { PopupService } from '../core/core.module';
export class Picker extends PopupService {
    /**
     * @param {?=} config
     * @param {?=} confirmCallback
     * @param {?=} cancelCallback
     * @return {?}
     */
    static showPicker(config = Picker.defaultOptions, confirmCallback, cancelCallback) {
        /** @type {?} */
        const options = new PickerOptions();
        Object.assign(options, config, {
            hidePicker: (event) => {
                this.hidePicker();
            },
            confirm: (event) => {
                if (confirmCallback) {
                    confirmCallback(event);
                }
            },
            cancel: () => {
                if (cancelCallback) {
                    cancelCallback();
                }
            }
        });
        /** @type {?} */
        const optionalParams = [
            'data',
            'value',
            'cols',
            'mask',
            'title',
            'okText',
            'dismissText',
            'disabled',
            'cascade',
            'onOk',
            'onPickerChange',
            'indicatorStyle'
        ];
        optionalParams.forEach(param => {
            if (typeof this[param] !== 'undefined') {
                (/** @type {?} */ (options))[param] = this[param];
            }
        });
        /** @type {?} */
        const childInjector = Injector.create([
            {
                provide: PickerOptions,
                useValue: options
            }
        ]);
        Picker.showPopup(PickerComponent, childInjector);
    }
    /**
     * @return {?}
     */
    static hidePicker() {
        Picker.hidePopup();
    }
}
Picker.defaultOptions = new PickerOptions();
Picker.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    Picker.defaultOptions;
}
