// ng-zorro-antd-mobile/esm2015/action-sheet/action-sheet-subject.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/** @enum {number} */
const actionSheetEvent = {
    onShow: 0,
    onShown: 1,
    onHide: 2,
    onHidden: 3,
    onOk: 4,
    onCancel: 5,
    onDestroy: 6,
};
export class ActionSheetSubject extends Subject {
    constructor() {
        super();
        this._eventsQueue = {};
        this.subscribe((value) => {
            /** @type {?} */
            const eventQueue = this._eventsQueue[value] || [];
            eventQueue.forEach(cb => {
                if (cb !== null && cb !== undefined) {
                    cb();
                }
            });
        });
    }
    /**
     * @param {?=} type
     * @return {?}
     */
    destroy(type = 'onCancel') {
        if (!this.isStopped && !this.closed) {
            this.next(type);
        }
    }
    /**
     * @param {?} eventType
     * @param {?} cb
     * @return {?}
     */
    on(eventType, cb) {
        if (this._eventsQueue[eventType]) {
            this._eventsQueue[eventType].push(cb);
        }
        else {
            this._eventsQueue[eventType] = [cb];
        }
    }
}
ActionSheetSubject.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ActionSheetSubject.ctorParameters = () => [];
if (false) {
    /** @type {?} */
    ActionSheetSubject.prototype.modalId;
    /** @type {?} */
    ActionSheetSubject.prototype._eventsQueue;
}
