// ng-zorro-antd-mobile/esm5/calendar/shortcut-panel/shortcut-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, HostBinding, Input } from '@angular/core';
/**
 * @record
 */
export function PropsType() { }
/** @type {?} */
PropsType.prototype.locale;
/** @type {?} */
PropsType.prototype.onSelect;
var ShortcutPanelComponent = /** @class */ (function () {
    function ShortcutPanelComponent() {
        var _this = this;
        this.props = /** @type {?} */ ({});
        this.shortcutPanel = true;
        this.onClick = function (type) {
            var onSelect = _this.props.onSelect;
            /** @type {?} */
            var today = new Date();
            switch (type) {
                case 'today':
                    onSelect(new Date(today.getFullYear(), today.getMonth(), today.getDate(), 0), new Date(today.getFullYear(), today.getMonth(), today.getDate(), 12));
                    break;
                case 'yesterday':
                    onSelect(new Date(today.getFullYear(), today.getMonth(), today.getDate() - 1, 0), new Date(today.getFullYear(), today.getMonth(), today.getDate() - 1, 12));
                    break;
                case 'lastweek':
                    onSelect(new Date(today.getFullYear(), today.getMonth(), today.getDate() - 6, 0), new Date(today.getFullYear(), today.getMonth(), today.getDate(), 12));
                    break;
                case 'lastmonth':
                    onSelect(new Date(today.getFullYear(), today.getMonth(), today.getDate() - 29, 0), new Date(today.getFullYear(), today.getMonth(), today.getDate(), 12));
                    break;
            }
        };
    }
    Object.defineProperty(ShortcutPanelComponent.prototype, "locale", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.locale = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ShortcutPanelComponent.prototype, "onSelect", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.props.onSelect = value;
        },
        enumerable: true,
        configurable: true
    });
    ShortcutPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'CalendarShortcutPanel, nzm-calendar-shortcut-panel',
                    template: "<div class=\"item\" (click)=\"onClick('today')\">{{props.locale.today}}</div>\n<div class=\"item\" (click)=\"onClick('yesterday')\">{{props.locale.yesterday}}</div>\n<div class=\"item\" (click)=\"onClick('lastweek')\">{{props.locale.lastWeek}}</div>\n<div class=\"item\" (click)=\"onClick('lastmonth')\">{{props.locale.lastMonth}}</div>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    ShortcutPanelComponent.ctorParameters = function () { return []; };
    ShortcutPanelComponent.propDecorators = {
        locale: [{ type: Input }],
        onSelect: [{ type: Input }],
        shortcutPanel: [{ type: HostBinding, args: ['class.shortcut-panel',] }]
    };
    return ShortcutPanelComponent;
}());
export { ShortcutPanelComponent };
if (false) {
    /** @type {?} */
    ShortcutPanelComponent.prototype.props;
    /** @type {?} */
    ShortcutPanelComponent.prototype.shortcutPanel;
    /** @type {?} */
    ShortcutPanelComponent.prototype.onClick;
}
