// ng-zorro-antd-mobile/esm5/toast/toast.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, ApplicationRef, Compiler, NgZone, ComponentFactoryResolver, } from '@angular/core';
import { ToastComponent } from './toast.component';
import { ToastOptions } from './toast-options.provider';
/**
 * @record
 */
export function ConfigInterface() { }
/** @type {?|undefined} */
ConfigInterface.prototype.content;
/** @type {?|undefined} */
ConfigInterface.prototype.iconType;
/** @type {?|undefined} */
ConfigInterface.prototype.mask;
var Toast = /** @class */ (function () {
    function Toast(_appRef, _compiler, _cfr, _zone) {
        this._appRef = _appRef;
        this._compiler = _compiler;
        this._cfr = _cfr;
        this._zone = _zone;
        Toast._zone = this._zone;
        Toast._appRef = this._appRef;
        Toast._toastCompFactory = this._cfr.resolveComponentFactory(ToastComponent);
    }
    /**
     * @param {?} config
     * @param {?} options
     * @return {?}
     */
    Toast._initConfig = /**
     * @param {?} config
     * @param {?} options
     * @return {?}
     */
    function (config, options) {
        /** @type {?} */
        var props = {};
        /** @type {?} */
        var optionalParams = ['content', 'iconType', 'mask'];
        config = Object.assign(options, config);
        optionalParams.forEach(function (key) {
            if (config[key] !== undefined) {
                props[key] = config[key];
            }
        });
        /** @type {?} */
        var iconType = {
            info: '',
            success: 'success',
            fail: 'fail',
            offline: 'dislike',
            loading: 'loading'
        }[options.iconType];
        props['iconType'] = iconType;
        props['mask'] = options.mask;
        return props;
    };
    /**
     * @param {?} config
     * @param {?} type
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    Toast.notice = /**
     * @param {?} config
     * @param {?} type
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    function (config, type, timeInterval, onClose, mask) {
        if (timeInterval === void 0) { timeInterval = 2000; }
        if (mask === void 0) { mask = true; }
        // 如果已经存在，在没有遮罩层的情况下，会响应别的toast，需要清除原来的
        if (Toast.compRef) {
            Toast.hide();
        }
        /** @type {?} */
        var options = new ToastOptions();
        options.iconType = type;
        options.mask = mask;
        /** @type {?} */
        var props = Toast._initConfig(config, options);
        document.body.insertBefore(document.createElement(Toast._toastCompFactory.selector), document.body.firstChild);
        /** @type {?} */
        var instance;
        /** @type {?} */
        var subject;
        Toast.compRef = Toast._appRef.bootstrap(Toast._toastCompFactory);
        instance = Toast.compRef.instance;
        subject = instance.subject;
        if (timeInterval) {
            Toast.timeout = setTimeout(function () {
                if (onClose) {
                    onClose();
                }
                Toast.hide();
            }, timeInterval);
        }
        Object.assign(instance, props);
        return subject;
    };
    /**
     * Open info dialog
     */
    /**
     * Open info dialog
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    Toast.info = /**
     * Open info dialog
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    function (content, timeInterval, onClose, mask) {
        /** @type {?} */
        var config = Object.assign({
            iconType: 'info',
            content: content
        });
        return Toast.notice(config, 'info', timeInterval, onClose, mask);
    };
    /**
     * Open success dialog
     */
    /**
     * Open success dialog
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    Toast.success = /**
     * Open success dialog
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    function (content, timeInterval, onClose, mask) {
        /** @type {?} */
        var config = Object.assign({
            iconType: 'success',
            content: content
        });
        return Toast.notice(config, 'success', timeInterval, onClose, mask);
    };
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} mask
     * @return {?}
     */
    Toast.show = /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} mask
     * @return {?}
     */
    function (content, timeInterval, mask) {
        /** @type {?} */
        var config = Object.assign({
            iconType: 'info',
            content: content
        });
        return Toast.notice(config, 'info', timeInterval, function () { }, mask);
    };
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    Toast.fail = /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    function (content, timeInterval, onClose, mask) {
        /** @type {?} */
        var config = Object.assign({
            iconType: 'fail',
            content: content
        });
        return Toast.notice(config, 'fail', timeInterval, onClose, mask);
    };
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    Toast.offline = /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    function (content, timeInterval, onClose, mask) {
        /** @type {?} */
        var config = Object.assign({
            iconType: 'offline',
            content: content
        });
        return Toast.notice(config, 'offline', timeInterval, onClose, mask);
    };
    /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    Toast.loading = /**
     * @param {?=} content
     * @param {?=} timeInterval
     * @param {?=} onClose
     * @param {?=} mask
     * @return {?}
     */
    function (content, timeInterval, onClose, mask) {
        /** @type {?} */
        var config = Object.assign({
            iconType: 'loading',
            content: content
        });
        return Toast.notice(config, 'loading', timeInterval, onClose, mask);
    };
    /**
     * @return {?}
     */
    Toast.hide = /**
     * @return {?}
     */
    function () {
        if (Toast.timeout) {
            clearTimeout(Toast.timeout);
        }
        if (Toast.compRef) {
            Toast._zone.run(function () {
                Toast.compRef.destroy();
            });
            Toast.compRef = null;
        }
    };
    Toast.timeout = null;
    Toast._zone = null;
    Toast.compRef = null;
    Toast._toastCompFactory = null;
    Toast._appRef = null;
    Toast.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    Toast.ctorParameters = function () { return [
        { type: ApplicationRef },
        { type: Compiler },
        { type: ComponentFactoryResolver },
        { type: NgZone }
    ]; };
    return Toast;
}());
export { Toast };
if (false) {
    /** @type {?} */
    Toast.timeout;
    /** @type {?} */
    Toast._zone;
    /** @type {?} */
    Toast.compRef;
    /** @type {?} */
    Toast._toastCompFactory;
    /** @type {?} */
    Toast._appRef;
    /** @type {?} */
    Toast.prototype._appRef;
    /** @type {?} */
    Toast.prototype._compiler;
    /** @type {?} */
    Toast.prototype._cfr;
    /** @type {?} */
    Toast.prototype._zone;
}
