// ng-zorro-antd-mobile/esm5/switch/switch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, EventEmitter, forwardRef, HostBinding } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var Switch = /** @class */ (function () {
    function Switch() {
        this.prefixCls = 'am-switch';
        this.wrapCls = 'am-switch';
        this.checkboxCls = {
            'checkbox-disabled': false
        };
        this.colorStyle = {};
        this._color = '';
        this._platform = 'ios';
        this._disabled = false;
        this.onChange = new EventEmitter();
        this.onClick = new EventEmitter();
        this.dispaly = true;
    }
    Object.defineProperty(Switch.prototype, "color", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._color = value;
            this.colorStyle = { background: this._color };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Switch.prototype, "platform", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._platform = value;
            this.wrapCls = value === 'android' ? this.prefixCls + "-android" : this.prefixCls;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Switch.prototype, "checked", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.switchChecked = value;
            this.colorStyle = { background: value ? this._color : '' };
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Switch.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
            this.checkboxCls = {
                'checkbox-disabled': value
            };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} checkedValue
     * @return {?}
     */
    Switch.prototype.changeSwitch = /**
     * @param {?} checkedValue
     * @return {?}
     */
    function (checkedValue) {
        this.switchChecked = checkedValue;
        this.colorStyle = { background: checkedValue ? this._color : '' };
        this.onChange.emit(checkedValue);
    };
    /**
     * @return {?}
     */
    Switch.prototype.click = /**
     * @return {?}
     */
    function () {
        this.onClick.emit(this.switchChecked);
    };
    Switch.decorators = [
        { type: Component, args: [{
                    selector: 'Switch, nzm-switch',
                    template: "<label class=\"{{prefixCls}}\" [ngClass]=\"wrapCls\">\n  <input #switchValue\n         type=\"checkbox\"\n         name=\"name\"\n         class=\"{{prefixCls}}-checkbox\"\n         [checked]=\"switchChecked\"\n         [value]=\"switchChecked\"\n         [disabled]=\"disabled\"\n         (change)=\"changeSwitch(switchValue.checked)\"\n  />\n  <div class=\"checkbox\"\n       [ngClass]=\"checkboxCls\"\n       [ngStyle]=\"colorStyle\"\n       (click)=\"click()\"\n  ></div>\n</label>",
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return Switch; }),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    Switch.ctorParameters = function () { return []; };
    Switch.propDecorators = {
        color: [{ type: Input }],
        name: [{ type: Input }],
        platform: [{ type: Input }],
        checked: [{ type: Input }],
        disabled: [{ type: Input }],
        onChange: [{ type: Output }],
        onClick: [{ type: Output }],
        dispaly: [{ type: HostBinding, args: ['style.display',] }]
    };
    return Switch;
}());
export { Switch };
if (false) {
    /** @type {?} */
    Switch.prototype.prefixCls;
    /** @type {?} */
    Switch.prototype.wrapCls;
    /** @type {?} */
    Switch.prototype.checkboxCls;
    /** @type {?} */
    Switch.prototype.colorStyle;
    /** @type {?} */
    Switch.prototype.switchChecked;
    /** @type {?} */
    Switch.prototype._color;
    /** @type {?} */
    Switch.prototype._platform;
    /** @type {?} */
    Switch.prototype._disabled;
    /** @type {?} */
    Switch.prototype.name;
    /** @type {?} */
    Switch.prototype.onChange;
    /** @type {?} */
    Switch.prototype.onClick;
    /** @type {?} */
    Switch.prototype.dispaly;
}
