// ng-zorro-antd-mobile/esm2015/input-item/custom-input/custom-input.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable, ApplicationRef, ComponentFactoryResolver } from '@angular/core';
import { CustomKeyboard } from '../custom-keyboard/custom-keyboard.component';
export class CustomInputService {
    /**
     * @param {?} _appRef
     * @param {?} _cfr
     */
    constructor(_appRef, _cfr) {
        this._appRef = _appRef;
        this._cfr = _cfr;
        CustomInputService.appRef = this._appRef;
        CustomInputService._inputCompFactory = this._cfr.resolveComponentFactory(CustomKeyboard);
    }
    /**
     * @return {?}
     */
    static getShowStatus() {
        return CustomInputService.isShow;
    }
    /**
     * @return {?}
     */
    static showKeyboard() {
        if (!this.isShow) {
            if (this.compRef) {
                this.compRef.instance.wrapperCls = {
                    [`am-number-keyboard-wrapper`]: true
                };
            }
            else {
                /** @type {?} */
                let container = document.querySelector(`#${this._keyboardPrefixCls}-container`);
                if (!container) {
                    container = document.createElement('div');
                    container.setAttribute('id', `${this._keyboardPrefixCls}-container`);
                    document.body.appendChild(container);
                    container.appendChild(document.createElement(CustomInputService._inputCompFactory.selector));
                    this.compRef = this.appRef.bootstrap(CustomInputService._inputCompFactory);
                    this.compRef.instance.onClick.subscribe(e => {
                        this.clickValue = e;
                    });
                }
            }
            this.isShow = true;
        }
    }
    /**
     * @return {?}
     */
    static hideKeyboard() {
        if (this.compRef && this.isShow) {
            this.isShow = false;
            this.compRef.instance.wrapperCls = {
                [`am-number-keyboard-wrapper`]: true,
                [`am-number-keyboard-wrapper-hide`]: true
            };
        }
    }
}
CustomInputService.compRef = null;
CustomInputService.appRef = null;
CustomInputService.isShow = false;
CustomInputService.clickValue = null;
CustomInputService._inputCompFactory = null;
CustomInputService._keyboardPrefixCls = 'am-number-keyboard';
CustomInputService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
CustomInputService.ctorParameters = () => [
    { type: ApplicationRef },
    { type: ComponentFactoryResolver }
];
if (false) {
    /** @type {?} */
    CustomInputService.compRef;
    /** @type {?} */
    CustomInputService.appRef;
    /** @type {?} */
    CustomInputService.isShow;
    /** @type {?} */
    CustomInputService.clickValue;
    /** @type {?} */
    CustomInputService._inputCompFactory;
    /** @type {?} */
    CustomInputService._keyboardPrefixCls;
    /** @type {?} */
    CustomInputService.prototype._appRef;
    /** @type {?} */
    CustomInputService.prototype._cfr;
}
