// ng-zorro-antd-mobile/esm5/input-item/custom-keyboard/custom-keyboard.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, ViewEncapsulation } from '@angular/core';
import { LocaleProviderService } from '../../locale-provider/locale-provider.service';
import { takeUntil } from 'rxjs/operators';
import { Subject } from 'rxjs';
var CustomKeyboard = /** @class */ (function () {
    function CustomKeyboard(_localeProvider) {
        this._localeProvider = _localeProvider;
        this.prefixCls = 'am-number-keyboard';
        this.okText = '';
        this._locale = {};
        this._unsubscribe$ = new Subject();
        this.onClick = new EventEmitter();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    CustomKeyboard.prototype.tdClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.onClick.emit(e);
    };
    /**
     * @return {?}
     */
    CustomKeyboard.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var _a, _b;
        this.wrapCls = (_a = {},
            _a[this.prefixCls + "-item"] = true,
            _a);
        this.wrapperCls = (_b = {},
            _b[this.prefixCls + "-wrapper"] = true,
            _b);
        this._localeProvider.localeChange.pipe(takeUntil(this._unsubscribe$)).subscribe(function (_) {
            _this._locale = _this._localeProvider.getLocaleSubObj('InputItem');
            _this.okText = _this._locale.confirmLabel;
        });
    };
    /**
     * @return {?}
     */
    CustomKeyboard.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._unsubscribe$.next();
        this._unsubscribe$.complete();
    };
    CustomKeyboard.decorators = [
        { type: Component, args: [{
                    selector: 'CustomKeyboard',
                    template: "<div [ngClass]=\"wrapperCls\">\n<table>\n  <tbody>\n    <tr>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(1)\">1\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(2)\">2\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(3)\">3\n      </td>\n      <td class=\"keyboard-delete\" [rowSpan]=\"2\" [ngClass]=\"wrapCls\" (click)=\"tdClick('delete')\">\n      </td>\n    </tr>\n    <tr>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(4)\">4\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(5)\">5\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(6)\">6\n      </td>\n\n    </tr>\n    <tr>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(7)\">7\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(8)\">8\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(9)\">9\n      </td>\n      <td class=\"keyboard-confirm\" [rowSpan]=\"2\" [ngClass]=\"wrapCls\" (click)=\"tdClick('confirm')\">{{okText}}\n      </td>\n    </tr>\n    <tr>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick('.')\">.\n      </td>\n      <td [ngClass]=\"wrapCls\" (click)=\"tdClick(0)\">0\n      </td>\n      <td class=\"keyboard-hide\" [ngClass]=\"wrapCls\" (click)=\"tdClick('hide')\">\n      </td>\n    </tr>\n  </tbody>\n</table>\n</div>\n\n",
                    encapsulation: ViewEncapsulation.None,
                    providers: [LocaleProviderService]
                }] }
    ];
    /** @nocollapse */
    CustomKeyboard.ctorParameters = function () { return [
        { type: LocaleProviderService }
    ]; };
    CustomKeyboard.propDecorators = {
        onClick: [{ type: Output }]
    };
    return CustomKeyboard;
}());
export { CustomKeyboard };
if (false) {
    /** @type {?} */
    CustomKeyboard.prototype.prefixCls;
    /** @type {?} */
    CustomKeyboard.prototype.wrapCls;
    /** @type {?} */
    CustomKeyboard.prototype.okText;
    /** @type {?} */
    CustomKeyboard.prototype.wrapperCls;
    /** @type {?} */
    CustomKeyboard.prototype._locale;
    /** @type {?} */
    CustomKeyboard.prototype._unsubscribe$;
    /** @type {?} */
    CustomKeyboard.prototype.onClick;
    /** @type {?} */
    CustomKeyboard.prototype._localeProvider;
}
