// ng-zorro-antd-mobile/esm5/picker/picker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter, HostListener, ViewContainerRef, ElementRef, Injector, ComponentFactoryResolver, Renderer2 } from '@angular/core';
import { PickerComponent } from './picker.component';
import { PickerOptions } from './picker-options.provider';
var PickerDirective = /** @class */ (function () {
    function PickerDirective(_viewContainerRef, _elm, _defaultOptions, _cfr, _renderer) {
        this._viewContainerRef = _viewContainerRef;
        this._elm = _elm;
        this._defaultOptions = _defaultOptions;
        this._cfr = _cfr;
        this._renderer = _renderer;
        this._eventListeners = [];
        this.onVisibleChange = new EventEmitter(true);
        this.onPickerChange = new EventEmitter();
        this.onOk = new EventEmitter();
        this.onDismiss = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PickerDirective.prototype.togglePicker = /**
     * @return {?}
     */
    function () {
        if (!this.picker) {
            this.showPicker();
        }
        else {
            this.hidePicker();
        }
    };
    /**
     * @return {?}
     */
    PickerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.onVisibleChange.emit(false);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    PickerDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes["isOpen"]) {
            if (changes["isOpen"].currentValue === true) {
                this.showPicker();
            }
            else {
                this.hidePicker();
            }
        }
    };
    /**
     * @return {?}
     */
    PickerDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.hidePicker();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PickerDirective.prototype.onDocumentClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.picker &&
            !this._elm.nativeElement.contains(event.target) &&
            !this.picker.location.nativeElement.contains(event.target)) {
            this.hidePicker();
        }
    };
    /**
     * @return {?}
     */
    PickerDirective.prototype.showPicker = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.picker) {
            setTimeout(function () {
                _this._eventListeners = [
                    _this._renderer.listen('document', 'click', function (event) { return _this.onDocumentClick(event); }),
                    _this._renderer.listen('document', 'touchend', function (event) { return _this.onDocumentClick(event); })
                ];
            });
            /** @type {?} */
            var options_1 = new PickerOptions();
            Object.assign(options_1, this._defaultOptions, {
                hidePicker: function (event) {
                    _this.hidePicker();
                }
            });
            /** @type {?} */
            var optionalParams = [
                'data',
                'value',
                'cols',
                'mask',
                'title',
                'okText',
                'dismissText',
                'disabled',
                'cascade',
                'appendToBody',
                'indicatorStyle',
                'onOk',
                'onPickerChange'
            ];
            optionalParams.forEach(function (param) {
                if (typeof _this[param] !== 'undefined') {
                    (/** @type {?} */ (options_1))[param] = _this[param];
                }
            });
            /** @type {?} */
            var componentFactory = this._cfr.resolveComponentFactory(PickerComponent);
            /** @type {?} */
            var childInjector = Injector.create([
                {
                    provide: PickerOptions,
                    useValue: options_1
                }
            ]);
            this.picker = this._viewContainerRef.createComponent(componentFactory, this._viewContainerRef.length, childInjector);
            if (options_1.appendToBody) {
                document.body.appendChild(this.picker.location.nativeElement);
            }
            this.onVisibleChange.emit(true);
        }
    };
    /**
     * @return {?}
     */
    PickerDirective.prototype.hidePicker = /**
     * @return {?}
     */
    function () {
        if (this.picker) {
            this.picker.destroy();
            delete this.picker;
            this.onVisibleChange.emit(false);
            this._eventListeners.forEach(function (fn) { return fn(); });
            this._eventListeners = [];
        }
    };
    PickerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[Picker], [nzm-picker]'
                },] }
    ];
    /** @nocollapse */
    PickerDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ElementRef },
        { type: PickerOptions },
        { type: ComponentFactoryResolver },
        { type: Renderer2 }
    ]; };
    PickerDirective.propDecorators = {
        data: [{ type: Input }],
        value: [{ type: Input }],
        cols: [{ type: Input }],
        mask: [{ type: Input }],
        title: [{ type: Input }],
        visible: [{ type: Input }],
        okText: [{ type: Input }],
        dismissText: [{ type: Input }],
        disabled: [{ type: Input }],
        cascade: [{ type: Input }],
        appendToBody: [{ type: Input }],
        indicatorStyle: [{ type: Input }],
        onVisibleChange: [{ type: Output }],
        onPickerChange: [{ type: Output }],
        onOk: [{ type: Output }],
        onDismiss: [{ type: Output }],
        togglePicker: [{ type: HostListener, args: ['click',] }]
    };
    return PickerDirective;
}());
export { PickerDirective };
if (false) {
    /** @type {?} */
    PickerDirective.prototype.picker;
    /** @type {?} */
    PickerDirective.prototype._eventListeners;
    /** @type {?} */
    PickerDirective.prototype.data;
    /** @type {?} */
    PickerDirective.prototype.value;
    /** @type {?} */
    PickerDirective.prototype.cols;
    /** @type {?} */
    PickerDirective.prototype.mask;
    /** @type {?} */
    PickerDirective.prototype.title;
    /** @type {?} */
    PickerDirective.prototype.visible;
    /** @type {?} */
    PickerDirective.prototype.okText;
    /** @type {?} */
    PickerDirective.prototype.dismissText;
    /** @type {?} */
    PickerDirective.prototype.disabled;
    /** @type {?} */
    PickerDirective.prototype.cascade;
    /** @type {?} */
    PickerDirective.prototype.appendToBody;
    /** @type {?} */
    PickerDirective.prototype.indicatorStyle;
    /** @type {?} */
    PickerDirective.prototype.onVisibleChange;
    /** @type {?} */
    PickerDirective.prototype.onPickerChange;
    /** @type {?} */
    PickerDirective.prototype.onOk;
    /** @type {?} */
    PickerDirective.prototype.onDismiss;
    /** @type {?} */
    PickerDirective.prototype._viewContainerRef;
    /** @type {?} */
    PickerDirective.prototype._elm;
    /** @type {?} */
    PickerDirective.prototype._defaultOptions;
    /** @type {?} */
    PickerDirective.prototype._cfr;
    /** @type {?} */
    PickerDirective.prototype._renderer;
}
