// ng-zorro-antd-mobile/esm5/picker/picker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { PickerComponent } from './picker.component';
import { PickerOptions } from './picker-options.provider';
import { PopupService } from '../core/core.module';
var Picker = /** @class */ (function (_super) {
    tslib_1.__extends(Picker, _super);
    function Picker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?=} config
     * @param {?=} confirmCallback
     * @param {?=} cancelCallback
     * @return {?}
     */
    Picker.showPicker = /**
     * @param {?=} config
     * @param {?=} confirmCallback
     * @param {?=} cancelCallback
     * @return {?}
     */
    function (config, confirmCallback, cancelCallback) {
        var _this = this;
        if (config === void 0) { config = Picker.defaultOptions; }
        /** @type {?} */
        var options = new PickerOptions();
        Object.assign(options, config, {
            hidePicker: function (event) {
                _this.hidePicker();
            },
            confirm: function (event) {
                if (confirmCallback) {
                    confirmCallback(event);
                }
            },
            cancel: function () {
                if (cancelCallback) {
                    cancelCallback();
                }
            }
        });
        /** @type {?} */
        var optionalParams = [
            'data',
            'value',
            'cols',
            'mask',
            'title',
            'okText',
            'dismissText',
            'disabled',
            'cascade',
            'onOk',
            'onPickerChange',
            'indicatorStyle'
        ];
        optionalParams.forEach(function (param) {
            if (typeof _this[param] !== 'undefined') {
                (/** @type {?} */ (options))[param] = _this[param];
            }
        });
        /** @type {?} */
        var childInjector = Injector.create([
            {
                provide: PickerOptions,
                useValue: options
            }
        ]);
        Picker.showPopup(PickerComponent, childInjector);
    };
    /**
     * @return {?}
     */
    Picker.hidePicker = /**
     * @return {?}
     */
    function () {
        Picker.hidePopup();
    };
    Picker.defaultOptions = new PickerOptions();
    Picker.decorators = [
        { type: Injectable }
    ];
    return Picker;
}(PopupService));
export { Picker };
if (false) {
    /** @type {?} */
    Picker.defaultOptions;
}
