// ng-zorro-antd-mobile/esm5/result/result.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding, TemplateRef, ViewEncapsulation } from '@angular/core';
var Result = /** @class */ (function () {
    function Result() {
        this.prefixCls = 'am-result';
        this.isTitleString = true;
        this.isMessageString = true;
        this.onButtonClick = new EventEmitter();
        this.role = 'alert';
        this.amResult = true;
    }
    Object.defineProperty(Result.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this.isTemplateRef(value)) {
                this.isTitleString = false;
            }
            else {
                this.isTitleString = true;
            }
            this._title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Result.prototype, "message", {
        get: /**
         * @return {?}
         */
        function () {
            return this._message;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (this.isTemplateRef(value)) {
                this.isMessageString = false;
            }
            else {
                this.isMessageString = true;
            }
            this._message = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    Result.prototype.buttonClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onButtonClick.emit(event);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    Result.prototype.isTemplateRef = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            return value instanceof TemplateRef;
        }
        return false;
    };
    Result.decorators = [
        { type: Component, args: [{
                    selector: 'Result, nzm-result',
                    template: "<div *ngIf=\"img\" class=\"{{prefixCls}}-pic\">\n  <ng-template [ngTemplateOutlet]=\"img\"></ng-template>\n</div>\n<div *ngIf=\"!img && imgUrl\"\n     class=\"{{prefixCls}}-pic\"\n     [ngStyle]=\"{ backgroundImage: 'url('+ imgUrl +')' }\"\n></div>\n<div *ngIf=\"title\" class=\"{{prefixCls}}-title\">\n  <ng-template *ngIf=\"!isTitleString\" [ngTemplateOutlet]=\"title\"></ng-template>\n  <span *ngIf=\"isTitleString\">{{title}}</span>\n</div>\n<div [ngClass]=\"prefixCls+'-message'\">\n  <ng-template *ngIf=\"!isMessageString\" [ngTemplateOutlet]=\"message\"></ng-template>\n  <span *ngIf=\"isMessageString\">{{message}}</span>\n</div>\n<div *ngIf=\"buttonText\" class=\"{{prefixCls}}-button\">\n  <a Button [type]=\"buttonType\" (click)=\"buttonClick($event)\">\n    {{buttonText}}\n  </a>\n</div>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    Result.ctorParameters = function () { return []; };
    Result.propDecorators = {
        title: [{ type: Input }],
        imgUrl: [{ type: Input }],
        buttonText: [{ type: Input }],
        buttonType: [{ type: Input }],
        img: [{ type: Input }],
        message: [{ type: Input }],
        onButtonClick: [{ type: Output }],
        role: [{ type: HostBinding, args: ['attr.role',] }],
        amResult: [{ type: HostBinding, args: ['class.am-result',] }]
    };
    return Result;
}());
export { Result };
if (false) {
    /** @type {?} */
    Result.prototype.prefixCls;
    /** @type {?} */
    Result.prototype.isTitleString;
    /** @type {?} */
    Result.prototype.isMessageString;
    /** @type {?} */
    Result.prototype._title;
    /** @type {?} */
    Result.prototype._message;
    /** @type {?} */
    Result.prototype.imgUrl;
    /** @type {?} */
    Result.prototype.buttonText;
    /** @type {?} */
    Result.prototype.buttonType;
    /** @type {?} */
    Result.prototype.img;
    /** @type {?} */
    Result.prototype.onButtonClick;
    /** @type {?} */
    Result.prototype.role;
    /** @type {?} */
    Result.prototype.amResult;
}
