// ng-zorro-antd-mobile/esm5/checkbox/checkbox-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
/**
 * @record
 */
export function OnChangeEvent() { }
/** @type {?} */
OnChangeEvent.prototype.name;
/** @type {?} */
OnChangeEvent.prototype.value;
/** @type {?} */
OnChangeEvent.prototype.checked;
var CheckboxItem = /** @class */ (function () {
    function CheckboxItem() {
        this.prefixCls = 'am-checkbox';
        this.checked = false;
        this.disabled = false;
        this.onChange = new EventEmitter();
        this.onClick = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    CheckboxItem.prototype.onCheckboxClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
        if (this.onClick.observers.length > 0) {
            this.onClick.emit(event);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CheckboxItem.prototype.change = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onChange.emit(event);
    };
    CheckboxItem.decorators = [
        { type: Component, args: [{
                    selector: 'CheckboxItem, nzm-checkbox-item',
                    template: "<ListItem [className]=\"'am-checkbox-item ' + (disabled ? 'am-checkbox-item-disabled' : '')\"\n          [disabled]=\"disabled\"\n          [thumb]=\"checkbox\"\n          (onClick)=\"onCheckboxClick($event)\"\n>\n  <ng-content></ng-content>\n</ListItem>\n<ng-template #checkbox>\n  <label Checkbox\n         [name]=\"name\"\n         [value]=\"value\"\n         [checked]=\"checked\"\n         [disabled]=\"disabled\"\n         (onChange)=\"change($event)\"\n  >\n  </label>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    CheckboxItem.ctorParameters = function () { return []; };
    CheckboxItem.propDecorators = {
        name: [{ type: Input }],
        value: [{ type: Input }],
        checked: [{ type: Input }],
        disabled: [{ type: Input }],
        onChange: [{ type: Output }],
        onClick: [{ type: Output }]
    };
    return CheckboxItem;
}());
export { CheckboxItem };
if (false) {
    /** @type {?} */
    CheckboxItem.prototype.prefixCls;
    /** @type {?} */
    CheckboxItem.prototype.name;
    /** @type {?} */
    CheckboxItem.prototype.value;
    /** @type {?} */
    CheckboxItem.prototype.checked;
    /** @type {?} */
    CheckboxItem.prototype.disabled;
    /** @type {?} */
    CheckboxItem.prototype.onChange;
    /** @type {?} */
    CheckboxItem.prototype.onClick;
}
