// ng-zorro-antd-mobile/esm5/checkbox/checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding, HostListener, ViewEncapsulation, forwardRef } from '@angular/core';
import { toBoolean } from '../core/util/convert';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * @record
 */
export function OnChangeEvent() { }
/** @type {?} */
OnChangeEvent.prototype.name;
/** @type {?} */
OnChangeEvent.prototype.value;
/** @type {?} */
OnChangeEvent.prototype.checked;
var Checkbox = /** @class */ (function () {
    function Checkbox() {
        var _a;
        this.prefixCls = 'am-checkbox';
        this.classMap = (_a = {},
            _a[this.prefixCls] = true,
            _a[this.prefixCls + "-checked"] = this.checked,
            _a[this.prefixCls + "-disabled"] = this.disabled,
            _a);
        this._checked = false;
        this._disabled = false;
        this._ngModelOnChange = Function.prototype;
        this._ngModelOnTouched = Function.prototype;
        this.onChange = new EventEmitter();
        this.checkBoxWrapper = true;
    }
    Object.defineProperty(Checkbox.prototype, "checked", {
        get: /**
         * @return {?}
         */
        function () {
            return this._checked;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!this._disabled) {
                this._checked = value;
                this.updateClassMap();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Checkbox.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = toBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    Checkbox.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        if (!this._disabled) {
            this.updateValue(!this._checked);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    Checkbox.prototype.updateValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._ngModelOnChange(value);
        this.onChange.emit({ name: this.name, value: this.value, checked: value });
        this.checked = value;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    Checkbox.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (null !== value) {
            this.checked = value;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    Checkbox.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    Checkbox.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._ngModelOnTouched = fn;
    };
    /**
     * @return {?}
     */
    Checkbox.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.updateClassMap();
    };
    /**
     * @return {?}
     */
    Checkbox.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.updateClassMap();
    };
    /**
     * @return {?}
     */
    Checkbox.prototype.updateClassMap = /**
     * @return {?}
     */
    function () {
        var _a;
        this.classMap = (_a = {},
            _a[this.prefixCls] = true,
            _a[this.prefixCls + "-checked"] = this.checked,
            _a[this.prefixCls + "-disabled"] = this.disabled,
            _a);
    };
    Checkbox.decorators = [
        { type: Component, args: [{
                    selector: '[Checkbox], [nzm-checkbox]',
                    template: "<span [ngClass]=\"classMap\">\n  <input type=\"checkbox\"\n         class=\"{{prefixCls}}-input\"\n         [attr.name]=\"name\"\n         [attr.value]=\"value\"\n         [checked]=\"checked\"\n  >\n  <span class=\"{{prefixCls}}-inner\"></span>\n</span>\n<ng-content></ng-content>\n",
                    preserveWhitespaces: false,
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return Checkbox; }),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    Checkbox.ctorParameters = function () { return []; };
    Checkbox.propDecorators = {
        name: [{ type: Input }],
        value: [{ type: Input }],
        checked: [{ type: Input }],
        disabled: [{ type: Input }],
        onChange: [{ type: Output }],
        checkBoxWrapper: [{ type: HostBinding, args: ['class.am-checkbox-wrapper',] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return Checkbox;
}());
export { Checkbox };
if (false) {
    /** @type {?} */
    Checkbox.prototype.prefixCls;
    /** @type {?} */
    Checkbox.prototype.classMap;
    /** @type {?} */
    Checkbox.prototype._checked;
    /** @type {?} */
    Checkbox.prototype._disabled;
    /** @type {?} */
    Checkbox.prototype._ngModelOnChange;
    /** @type {?} */
    Checkbox.prototype._ngModelOnTouched;
    /** @type {?} */
    Checkbox.prototype.name;
    /** @type {?} */
    Checkbox.prototype.value;
    /** @type {?} */
    Checkbox.prototype.onChange;
    /** @type {?} */
    Checkbox.prototype.checkBoxWrapper;
}
