// ng-zorro-antd-mobile/esm2015/popover/popover-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PopoverOptionsInterface() { }
/** @type {?|undefined} */
PopoverOptionsInterface.prototype.mask;
/** @type {?|undefined} */
PopoverOptionsInterface.prototype.visible;
/** @type {?|undefined} */
PopoverOptionsInterface.prototype.placement;
/** @type {?|undefined} */
PopoverOptionsInterface.prototype.appendToBody;
/** @type {?|undefined} */
PopoverOptionsInterface.prototype.className;
/** @type {?|undefined} */
PopoverOptionsInterface.prototype.autoClose;
export class PopoverOptions {
    constructor() {
        this.showArrow = false;
        this.mask = false;
        this.placement = 'bottom';
        this.appendToBody = false;
        this.className = '';
        this.autoClose = true;
    }
}
if (false) {
    /** @type {?} */
    PopoverOptions.prototype.showArrow;
    /** @type {?} */
    PopoverOptions.prototype.mask;
    /** @type {?} */
    PopoverOptions.prototype.placement;
    /** @type {?} */
    PopoverOptions.prototype.appendToBody;
    /** @type {?} */
    PopoverOptions.prototype.className;
    /** @type {?} */
    PopoverOptions.prototype.autoClose;
}
