// ng-zorro-antd-mobile/esm5/date-picker/date-picker.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DatePickerComponent } from './date-picker.component';
import { DatePickerDirective } from './date-picker.directive';
import { DatePickerOptions } from './date-picker-options.provider';
import { LocaleProviderModule } from '../locale-provider/locale-provider.module';
var DatePickerModule = /** @class */ (function () {
    function DatePickerModule() {
    }
    DatePickerModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, LocaleProviderModule],
                    exports: [DatePickerComponent, DatePickerDirective],
                    declarations: [DatePickerComponent, DatePickerDirective],
                    entryComponents: [DatePickerComponent],
                    providers: [DatePickerOptions, LocaleProviderModule]
                },] }
    ];
    return DatePickerModule;
}());
export { DatePickerModule };
