// ng-zorro-antd-mobile/esm5/core/core.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TouchFeedbackDirective } from './directive/touch-feedback.directive';
import { TouchFeedBackComponent } from './directive/touch-feedback.component';
import { PopupService } from './services/popup.service';
export { PopupService } from './services/popup.service';
import { OverlayModule } from '@angular/cdk/overlay';
var CoreModule = /** @class */ (function () {
    function CoreModule() {
    }
    CoreModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        OverlayModule
                    ],
                    exports: [TouchFeedbackDirective, TouchFeedBackComponent],
                    declarations: [TouchFeedbackDirective, TouchFeedBackComponent],
                    providers: [PopupService]
                },] }
    ];
    return CoreModule;
}());
export { CoreModule };
