// ng-zorro-antd-mobile/esm2015/modal/modal-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class BaseOptions {
    constructor() {
        this.visible = false;
        this.focus = true;
        this.closable = false;
        this.maskClosable = true;
        this.transparent = false;
        this.popup = false;
        this.transitionName = 'am-zoom';
        this.maskTransitionName = 'am-fade';
    }
}
if (false) {
    /** @type {?} */
    BaseOptions.prototype.visible;
    /** @type {?} */
    BaseOptions.prototype.focus;
    /** @type {?} */
    BaseOptions.prototype.closable;
    /** @type {?} */
    BaseOptions.prototype.maskClosable;
    /** @type {?} */
    BaseOptions.prototype.onClose;
    /** @type {?} */
    BaseOptions.prototype.transparent;
    /** @type {?} */
    BaseOptions.prototype.popup;
    /** @type {?} */
    BaseOptions.prototype.animationType;
    /** @type {?} */
    BaseOptions.prototype.title;
    /** @type {?} */
    BaseOptions.prototype.footer;
    /** @type {?} */
    BaseOptions.prototype.platform;
    /** @type {?} */
    BaseOptions.prototype.className;
    /** @type {?} */
    BaseOptions.prototype.wrapClassName;
    /** @type {?} */
    BaseOptions.prototype.message;
    /** @type {?} */
    BaseOptions.prototype.actions;
    /** @type {?} */
    BaseOptions.prototype.callbackOrActions;
    /** @type {?} */
    BaseOptions.prototype.type;
    /** @type {?} */
    BaseOptions.prototype.defaultValue;
    /** @type {?} */
    BaseOptions.prototype.placeholders;
    /** @type {?} */
    BaseOptions.prototype.operation;
    /** @type {?} */
    BaseOptions.prototype.transitionName;
    /** @type {?} */
    BaseOptions.prototype.maskTransitionName;
    /** @type {?} */
    BaseOptions.prototype.close;
}
export class ModalOptions extends BaseOptions {
}
ModalOptions.decorators = [
    { type: Injectable }
];
export class AlertOptions extends BaseOptions {
}
AlertOptions.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * 	提示信息
     * @type {?}
     */
    AlertOptions.prototype.message;
    /**
     * 按钮组, [{text, onPress, style}]
     * @type {?}
     */
    AlertOptions.prototype.actions;
}
export class PromptOptions extends BaseOptions {
}
PromptOptions.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * 	提示信息
     * @type {?}
     */
    PromptOptions.prototype.message;
    /**
     * 按钮组 [{text, onPress}]
     * 或回调函数 Array or Function
     * @type {?}
     */
    PromptOptions.prototype.callbackOrActions;
    /**
     * prompt 的样式
     * String (default, secure-text, login-password)
     * @type {?}
     */
    PromptOptions.prototype.type;
    /**
     * 默认值(input 为 password 类型不支持)
     * @type {?}
     */
    PromptOptions.prototype.defaultValue;
    /**
     * ['', '']
     * @type {?}
     */
    PromptOptions.prototype.placeholders;
}
export class OperationOptions extends BaseOptions {
}
OperationOptions.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * 按钮组, [{text, onPress, style}]
     * @type {?}
     */
    OperationOptions.prototype.actions;
}
export class Action {
}
Action.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    Action.prototype.text;
    /** @type {?} */
    Action.prototype.onPress;
    /** @type {?} */
    Action.prototype.style;
}
