// ng-zorro-antd-mobile/esm5/nav-bar/nav-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef, Output, EventEmitter, HostBinding } from '@angular/core';
var NavBarComponent = /** @class */ (function () {
    function NavBarComponent() {
        this.defaultProps = {
            prefixCls: 'am-navbar',
            mode: 'dark',
            onLeftClick: function () { }
        };
        this.navbarCls = {};
        this.isIconString = true;
        this.isLeftContentString = true;
        this.isRightContentString = true;
        this.onLeftClick = new EventEmitter();
        this._amNavbar = true;
    }
    Object.defineProperty(NavBarComponent.prototype, "mode", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.defaultProps.mode = value;
            this._amNavbarLight = this.defaultProps.mode === 'light';
            this._amNavbardark = this.defaultProps.mode === 'dark';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavBarComponent.prototype, "icon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._icon;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this.isIconString = false;
            }
            else {
                this.isIconString = true;
            }
            this._icon = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavBarComponent.prototype, "leftContent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._leftContent;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this.isLeftContentString = false;
            }
            else {
                this.isLeftContentString = true;
            }
            this._leftContent = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NavBarComponent.prototype, "rightContent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rightContent;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this.isRightContentString = false;
            }
            else {
                this.isRightContentString = true;
            }
            this._rightContent = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    NavBarComponent.prototype.click = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onLeftClick.emit(event);
    };
    NavBarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'Navbar, nzm-nav-bar',
                    template: "<div role=\"button\" class=\"{{defaultProps.prefixCls}}-left\" (click)=\"click($event)\">\n  <ng-template *ngIf=\"!isLeftContentString\" [ngTemplateOutlet]=\"leftContent\"></ng-template>\n  <span *ngIf=\"icon\" class=\"{{defaultProps.prefixCls}}-left-icon\" aria-hidden=\"true\">\n    <Icon *ngIf=\"isIconString\" [type]=\"icon\"></Icon>\n    <ng-template *ngIf=\"!isIconString\" [ngTemplateOutlet]=\"icon\"></ng-template>\n  </span>\n  {{isLeftContentString ? leftContent: null}}\n</div>\n<div class=\"{{defaultProps.prefixCls}}-title\">\n  <ng-content></ng-content>\n</div>\n<div class=\"{{defaultProps.prefixCls}}-right\">\n  {{isRightContentString ? rightContent: null}}\n  <ng-template *ngIf=\"!isRightContentString\" [ngTemplateOutlet]=\"rightContent\"></ng-template>\n</div>"
                }] }
    ];
    /** @nocollapse */
    NavBarComponent.ctorParameters = function () { return []; };
    NavBarComponent.propDecorators = {
        mode: [{ type: Input }],
        icon: [{ type: Input }],
        leftContent: [{ type: Input }],
        rightContent: [{ type: Input }],
        onLeftClick: [{ type: Output }],
        _amNavbar: [{ type: HostBinding, args: ['class.am-navbar',] }],
        _amNavbarLight: [{ type: HostBinding, args: ['class.am-navbar-light',] }],
        _amNavbardark: [{ type: HostBinding, args: ['class.am-navbar-dark',] }]
    };
    return NavBarComponent;
}());
export { NavBarComponent };
if (false) {
    /** @type {?} */
    NavBarComponent.prototype.defaultProps;
    /** @type {?} */
    NavBarComponent.prototype.navbarCls;
    /** @type {?} */
    NavBarComponent.prototype.isIconString;
    /** @type {?} */
    NavBarComponent.prototype.isLeftContentString;
    /** @type {?} */
    NavBarComponent.prototype.isRightContentString;
    /** @type {?} */
    NavBarComponent.prototype._icon;
    /** @type {?} */
    NavBarComponent.prototype._leftContent;
    /** @type {?} */
    NavBarComponent.prototype._rightContent;
    /** @type {?} */
    NavBarComponent.prototype.onLeftClick;
    /** @type {?} */
    NavBarComponent.prototype._amNavbar;
    /** @type {?} */
    NavBarComponent.prototype._amNavbarLight;
    /** @type {?} */
    NavBarComponent.prototype._amNavbardark;
}
