// ng-zorro-antd-mobile/esm2015/calendar/confirm-panel/confirm-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
import { formatDate } from '../util/index';
/**
 * @record
 */
export function ConfirmPanelPropsType() { }
/** @type {?|undefined} */
ConfirmPanelPropsType.prototype.type;
/** @type {?} */
ConfirmPanelPropsType.prototype.locale;
/** @type {?|undefined} */
ConfirmPanelPropsType.prototype.onlyConfirm;
/** @type {?|undefined} */
ConfirmPanelPropsType.prototype.disableBtn;
/** @type {?|undefined} */
ConfirmPanelPropsType.prototype.startDateTime;
/** @type {?|undefined} */
ConfirmPanelPropsType.prototype.endDateTime;
/** @type {?|undefined} */
ConfirmPanelPropsType.prototype.formatStr;
/** @type {?} */
ConfirmPanelPropsType.prototype.onConfirm;
export class ConfirmPanelComponent {
    constructor() {
        this.props = /** @type {?} */ ({
            formatStr: 'yyyy-MM-dd hh:mm'
        });
        this.confirmPane = true;
        this.triggerConfirm = () => {
            const { onConfirm, disableBtn } = this.props;
            !disableBtn && onConfirm();
        };
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set propsData(value) {
        this.props = Object.assign({}, this.props, value);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set disableBtn(value) {
        this.props.disableBtn = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set formatStr(value) {
        this.props.formatStr = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set startDateTime(value) {
        this.props.startDateTime = value;
        this.formatTime();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set endDateTime(value) {
        this.props.endDateTime = value;
        this.formatTime();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set onConfirm(value) {
        this.props.onConfirm = value;
    }
    /**
     * @return {?}
     */
    formatTime() {
        const { type, locale, disableBtn } = this.props;
        let { startDateTime, endDateTime } = this.props;
        if (startDateTime && endDateTime && +startDateTime > +endDateTime) {
            /** @type {?} */
            const tmp = startDateTime;
            startDateTime = endDateTime;
            endDateTime = tmp;
        }
        this.startTimeStr = startDateTime ? this.selfFormatDate(startDateTime) : locale.noChoose;
        this.endTimeStr = endDateTime ? this.selfFormatDate(endDateTime) : locale.noChoose;
        /** @type {?} */
        let btnCls = disableBtn ? 'button button-disable' : 'button';
        if (type === 'one') {
            btnCls += ' button-full';
        }
        this.btnCls = btnCls;
    }
    /**
     * @param {?} date
     * @return {?}
     */
    selfFormatDate(date) {
        const { formatStr = '', locale } = this.props;
        return formatDate(date, formatStr, locale);
    }
}
ConfirmPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'CalendarConfirmPanel, nzm-calendar-confirm-panel',
                template: "<div *ngIf=\"props.type === 'range'\" class=\"info\">\n  <p>{{props.locale.start}}: <span class=\"{{!props.startDateTime ? 'grey' : ''}}\">{{startTimeStr}}</span></p>\n  <p>{{props.locale.end}}: <span class=\"{{!props.endDateTime ? 'grey' : ''}}\">{{endTimeStr}}</span></p>\n</div>\n<div [ngClass]=\"btnCls\" (click)=\"triggerConfirm()\">\n  {{props.locale.confirm}}\n</div>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
ConfirmPanelComponent.ctorParameters = () => [];
ConfirmPanelComponent.propDecorators = {
    propsData: [{ type: Input }],
    disableBtn: [{ type: Input }],
    formatStr: [{ type: Input }],
    startDateTime: [{ type: Input }],
    endDateTime: [{ type: Input }],
    onConfirm: [{ type: Input }],
    confirmPane: [{ type: HostBinding, args: ['class.confirm-panel',] }]
};
if (false) {
    /** @type {?} */
    ConfirmPanelComponent.prototype.props;
    /** @type {?} */
    ConfirmPanelComponent.prototype.startTimeStr;
    /** @type {?} */
    ConfirmPanelComponent.prototype.endTimeStr;
    /** @type {?} */
    ConfirmPanelComponent.prototype.btnCls;
    /** @type {?} */
    ConfirmPanelComponent.prototype.confirmPane;
    /** @type {?} */
    ConfirmPanelComponent.prototype.triggerConfirm;
}
