// ng-zorro-antd-mobile/esm2015/slider/slider-track/slider-track.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
export class SliderTrack {
    /**
     * @param {?} _elf
     * @param {?} _sanitizer
     */
    constructor(_elf, _sanitizer) {
        this._elf = _elf;
        this._sanitizer = _sanitizer;
        this.prefixCls = 'am-slider';
        this._included = true;
    }
    /**
     * @return {?}
     */
    get className() {
        return this._className;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set className(value) {
        this._className = value;
    }
    /**
     * @return {?}
     */
    get included() {
        return this._included;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set included(value) {
        this._included = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set offset(value) {
        this._offset = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set length(value) {
        this._length = value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set style(value) {
        this._style = value;
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        /** @type {?} */
        const positonStyle = {
            left: `${this._offset}%`,
            width: `${this._length}%`
        };
        this.elStyle = Object.assign({}, this._style, positonStyle);
    }
}
SliderTrack.decorators = [
    { type: Component, args: [{
                selector: 'SliderTrack, nzm-slider-track',
                template: "<div *ngIf=\"included\" \n     [ngClass]=\"className\" \n     [ngStyle]=\"elStyle\" >\n</div>",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
SliderTrack.ctorParameters = () => [
    { type: ElementRef },
    { type: DomSanitizer }
];
SliderTrack.propDecorators = {
    className: [{ type: Input }],
    included: [{ type: Input }],
    offset: [{ type: Input }],
    length: [{ type: Input }],
    style: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SliderTrack.prototype.prefixCls;
    /** @type {?} */
    SliderTrack.prototype.elStyle;
    /** @type {?} */
    SliderTrack.prototype._className;
    /** @type {?} */
    SliderTrack.prototype._included;
    /** @type {?} */
    SliderTrack.prototype._offset;
    /** @type {?} */
    SliderTrack.prototype._length;
    /** @type {?} */
    SliderTrack.prototype._style;
    /** @type {?} */
    SliderTrack.prototype._elf;
    /** @type {?} */
    SliderTrack.prototype._sanitizer;
}
