// ng-zorro-antd-mobile/esm2015/radio/radio-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
/**
 * @record
 */
export function OnChangeEvent() { }
/** @type {?} */
OnChangeEvent.prototype.name;
/** @type {?} */
OnChangeEvent.prototype.value;
/** @type {?} */
OnChangeEvent.prototype.checked;
export class RadioItem {
    constructor() {
        this.prefixCls = 'am-radio';
        this.checked = false;
        this.disabled = false;
        this.onClick = new EventEmitter();
        this.onChange = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onRadioItemClick(event) {
        if (!this.disabled && !this.checked) {
            this.checked = true;
            this.onChange.emit({
                name: this.name,
                value: this.value,
                checked: true
            });
        }
        if (this.onClick.observers.length > 0) {
            this.onClick.emit(event);
        }
    }
}
RadioItem.decorators = [
    { type: Component, args: [{
                selector: 'RadioItem, nzm-radio-item',
                template: "<ListItem [className]=\"'am-radio-item ' + (disabled ? 'am-radio-item-disabled' : '')\"\n          [extra]=\"radio\"\n          [disabled]=\"disabled\"\n          (onClick)=\"onRadioItemClick($event)\"\n>\n  <ng-content></ng-content>\n</ListItem>\n<ng-template #radio>\n  <label Radio\n         [name]=\"name\"\n         [value]=\"value\"\n         [checked]=\"checked\"\n         [disabled]=\"disabled\"\n  ></label>\n</ng-template>\n"
            }] }
];
/** @nocollapse */
RadioItem.ctorParameters = () => [];
RadioItem.propDecorators = {
    name: [{ type: Input }],
    value: [{ type: Input }],
    checked: [{ type: Input }],
    disabled: [{ type: Input }],
    onClick: [{ type: Output }],
    onChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    RadioItem.prototype.prefixCls;
    /** @type {?} */
    RadioItem.prototype.name;
    /** @type {?} */
    RadioItem.prototype.value;
    /** @type {?} */
    RadioItem.prototype.checked;
    /** @type {?} */
    RadioItem.prototype.disabled;
    /** @type {?} */
    RadioItem.prototype.onClick;
    /** @type {?} */
    RadioItem.prototype.onChange;
}
