// ng-zorro-antd-mobile/esm5/wing-blank/wing-blank.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, ViewEncapsulation } from '@angular/core';
/** @typedef {?} */
var sizeType;
export { sizeType };
var WingBlank = /** @class */ (function () {
    function WingBlank() {
        this.prefixCls = 'am-wingblank';
        this.size = 'lg';
        this.amWingBlank = true;
    }
    Object.defineProperty(WingBlank.prototype, "amWingBlnkSm", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'sm';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WingBlank.prototype, "amWingBlnkMd", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'md';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WingBlank.prototype, "amWingBlnkLg", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size === 'lg';
        },
        enumerable: true,
        configurable: true
    });
    WingBlank.decorators = [
        { type: Component, args: [{
                    selector: 'WingBlank, nzm-wingblank',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    WingBlank.ctorParameters = function () { return []; };
    WingBlank.propDecorators = {
        size: [{ type: Input }],
        amWingBlank: [{ type: HostBinding, args: ['class.am-wingblank',] }],
        amWingBlnkSm: [{ type: HostBinding, args: ['class.am-wingblank-sm',] }],
        amWingBlnkMd: [{ type: HostBinding, args: ['class.am-wingblank-md',] }],
        amWingBlnkLg: [{ type: HostBinding, args: ['class.am-wingblank-lg',] }]
    };
    return WingBlank;
}());
export { WingBlank };
if (false) {
    /** @type {?} */
    WingBlank.prototype.prefixCls;
    /** @type {?} */
    WingBlank.prototype.size;
    /** @type {?} */
    WingBlank.prototype.amWingBlank;
}
