// ng-zorro-antd-mobile/esm2015/steps/step/step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input, ElementRef, ViewEncapsulation, HostBinding } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
export class Step {
    /**
     * @param {?} _el
     */
    constructor(_el) {
        this._el = _el;
        this.prefixCls = 'am-steps';
        this.stepItemCls = {};
        this.isIconString = true;
        this.clsStepItem = true;
    }
    /**
     * @return {?}
     */
    get status() {
        return this._status;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set status(value) {
        if (value) {
            this._status = value;
            this.setClass();
        }
    }
    /**
     * @return {?}
     */
    get title() {
        return this._title;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set title(value) {
        this._title = value;
    }
    /**
     * @return {?}
     */
    get description() {
        return this._description;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set description(value) {
        this._description = value;
    }
    /**
     * @return {?}
     */
    get icon() {
        return this._icon;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set icon(value) {
        if (value) {
            if (typeof value !== 'string') {
                this.isIconString = false;
            }
            this._icon = value;
            this.setClass();
        }
    }
    /**
     * @return {?}
     */
    setClass() {
        this.iconCls = {
            [`${this.prefixCls}-icon`]: true
        };
        this.stepItemCls = Object.assign(this.stepItemCls, {
            [`${this.prefixCls}-item-${this.status}`]: true,
            [`${this.prefixCls}-item-custom`]: this.icon
        });
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
}
Step.decorators = [
    { type: Component, args: [{
                selector: 'Step, nzm-step',
                template: "<div [ngClass]=\"stepItemCls\">\n  <div class=\"{{prefixCls}}-item-tail\">\n    {{tailContent}}\n  </div>\n  <div class=\"{{prefixCls}}-item-icon\">\n    <span *ngIf=\"!isIconString\" [ngClass]=\"iconCls\">\n      <ng-template [ngTemplateOutlet]=\"icon\"></ng-template>\n    </span>\n    <span *ngIf=\"isIconString && (status=== 'error' || status==='finish' || status==='wait' )\" [ngClass]=\"iconCls\">\n      <Icon [type]=\"icon\" [size]=\"iconSize\">\n      </Icon>\n    </span>\n    <span *ngIf=\"isIconString && !(status=== 'error' || status==='finish' || status==='wait')\" [ngClass]=\"iconCls\">{{stepNumber}}\n    </span>\n  </div>\n  <div class=\"{{prefixCls}}-item-content\">\n    <div class=\"{{prefixCls}}-item-title\">\n      {{title}}\n    </div>\n    <div *ngIf=\"description\" class=\"{{prefixCls}}-item-description\">{{description}}</div>\n  </div>\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => Step),
                        multi: true
                    }
                ]
            }] }
];
/** @nocollapse */
Step.ctorParameters = () => [
    { type: ElementRef }
];
Step.propDecorators = {
    status: [{ type: Input }],
    title: [{ type: Input }],
    description: [{ type: Input }],
    icon: [{ type: Input }],
    clsStepItem: [{ type: HostBinding, args: ['class.am-steps-item',] }]
};
if (false) {
    /** @type {?} */
    Step.prototype.prefixCls;
    /** @type {?} */
    Step.prototype.stepItemCls;
    /** @type {?} */
    Step.prototype.iconCls;
    /** @type {?} */
    Step.prototype.tailContent;
    /** @type {?} */
    Step.prototype.stepNumber;
    /** @type {?} */
    Step.prototype.isIconString;
    /** @type {?} */
    Step.prototype._status;
    /** @type {?} */
    Step.prototype._title;
    /** @type {?} */
    Step.prototype._description;
    /** @type {?} */
    Step.prototype._icon;
    /** @type {?} */
    Step.prototype.clsStepItem;
    /** @type {?} */
    Step.prototype._el;
}
/** @enum {string} */
const StatusEnum = {
    WAIT: 'wait',
    PROCESS: 'process',
    FINISH: 'finish',
    ERROR: 'error',
};
export { StatusEnum };
/** @enum {string} */
const DirectionEnum = {
    VERTICAL: 'vertical',
    HORIZONTAL: 'horizontal',
};
export { DirectionEnum };
