// ng-zorro-antd-mobile/esm5/tab-bar/tab-bar-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostBinding, EventEmitter, Input, Output, TemplateRef, Optional } from '@angular/core';
import { TabPane } from '../tabs/tab-pane.component';
/**
 * @record
 */
export function OnPressEvent() { }
/** @type {?} */
OnPressEvent.prototype.title;
/** @type {?} */
OnPressEvent.prototype.key;
var TabBarTab = /** @class */ (function () {
    function TabBarTab(tabPane) {
        this.tabPane = tabPane;
        this.prefixCls = 'am-tab-bar-tab';
        this.selected = false;
        this.tintColor = '#108ee9';
        this.unselectedTintColor = '#888';
        this.key = '';
        this.title = '';
        this.dot = false;
        this.badge = null;
        this.icon = null;
        this.selectedIcon = null;
        this.onPress = new EventEmitter();
        this.tabBarTab = true;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    TabBarTab.prototype.isTemplateRef = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return value instanceof TemplateRef;
    };
    /**
     * @return {?}
     */
    TabBarTab.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        if (this.tabPane.active) {
            this.selected = true;
        }
        else {
            this.selected = false;
        }
        this.tintColor = this.tabPane.tintColor;
        this.unselectedTintColor = this.tabPane.unselectedTintColor;
    };
    TabBarTab.decorators = [
        { type: Component, args: [{
                    selector: 'TabBarTab, nzm-tab-bar-tab',
                    template: "<div class=\"{{prefixCls}}-icon\" [style.color]=\"selected ? tintColor : unselectedTintColor\">\n  <Badge *ngIf=\"badge\" className=\"{{prefixCls}}-badge tab-badge\" [text]=\"badge\">\n    <ng-container *ngIf=\"isTemplateRef(selected ? selectedIcon : icon); then domTemplate; else imgTemplate\">\n    </ng-container>\n  </Badge>\n  <Badge className=\"{{prefixCls}}-badge tab-badge\" [dot]=\"dot\" *ngIf=\"dot\">\n    <ng-container *ngIf=\"isTemplateRef(selected ? selectedIcon : icon); then domTemplate; else imgTemplate\">\n    </ng-container>\n  </Badge>\n  <ng-container *ngIf=\"!badge && !dot\" >\n    <ng-container *ngIf=\"isTemplateRef(selected ? selectedIcon : icon); then domTemplate; else imgTemplate\">\n    </ng-container>\n  </ng-container>\n</div>\n<p class=\"{{prefixCls}}-title\" [style.color]=\"selected ? tintColor : unselectedTintColor\">\n  {{title}}\n</p>\n\n<ng-template #domTemplate>\n  <ng-template [ngTemplateOutlet]=\"(selected ? selectedIcon : icon)\"></ng-template>\n</ng-template>\n<ng-template #imgTemplate >\n  <img src=\"{{selected ? selectedIcon : icon}}\" alt=\"{{title}}\">\n</ng-template>"
                }] }
    ];
    /** @nocollapse */
    TabBarTab.ctorParameters = function () { return [
        { type: TabPane, decorators: [{ type: Optional }] }
    ]; };
    TabBarTab.propDecorators = {
        key: [{ type: Input }],
        title: [{ type: Input }],
        dot: [{ type: Input }],
        badge: [{ type: Input }],
        icon: [{ type: Input }],
        selectedIcon: [{ type: Input }],
        onPress: [{ type: Output }],
        tabBarTab: [{ type: HostBinding, args: ['class.am-tab-bar-tab',] }]
    };
    return TabBarTab;
}());
export { TabBarTab };
if (false) {
    /** @type {?} */
    TabBarTab.prototype.prefixCls;
    /** @type {?} */
    TabBarTab.prototype.selected;
    /** @type {?} */
    TabBarTab.prototype.tintColor;
    /** @type {?} */
    TabBarTab.prototype.unselectedTintColor;
    /** @type {?} */
    TabBarTab.prototype.key;
    /** @type {?} */
    TabBarTab.prototype.title;
    /** @type {?} */
    TabBarTab.prototype.dot;
    /** @type {?} */
    TabBarTab.prototype.badge;
    /** @type {?} */
    TabBarTab.prototype.icon;
    /** @type {?} */
    TabBarTab.prototype.selectedIcon;
    /** @type {?} */
    TabBarTab.prototype.onPress;
    /** @type {?} */
    TabBarTab.prototype.tabBarTab;
    /** @type {?} */
    TabBarTab.prototype.tabPane;
}
