// ng-zorro-antd-mobile/esm2015/checkbox/agree-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, HostBinding } from '@angular/core';
/**
 * @record
 */
export function OnChangeEvent() { }
/** @type {?} */
OnChangeEvent.prototype.name;
/** @type {?} */
OnChangeEvent.prototype.value;
/** @type {?} */
OnChangeEvent.prototype.checked;
export class AgreeItem {
    constructor() {
        this.prefixCls = 'am-checkbox';
        this.checked = false;
        this.disabled = false;
        this.onChange = new EventEmitter();
        this.checkboxAgree = true;
    }
    /**
     * @return {?}
     */
    agree() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    change(event) {
        this.onChange.emit(event);
    }
}
AgreeItem.decorators = [
    { type: Component, args: [{
                selector: 'AgreeItem, nzm-agree-item',
                template: "<label Checkbox\n       class=\"{{prefixCls}}-agree-label\"\n       [name]=\"name\"\n       [value]=\"value\"\n       [checked]=\"checked\"\n       [disabled]=\"disabled\"\n       (click)=\"agree()\"\n       (onChange)=\"change($event)\"\n>\n  <ng-content></ng-content>\n</label>"
            }] }
];
/** @nocollapse */
AgreeItem.ctorParameters = () => [];
AgreeItem.propDecorators = {
    name: [{ type: Input }],
    value: [{ type: Input }],
    checked: [{ type: Input }],
    disabled: [{ type: Input }],
    onChange: [{ type: Output }],
    checkboxAgree: [{ type: HostBinding, args: ['class.am-checkbox-agree',] }]
};
if (false) {
    /** @type {?} */
    AgreeItem.prototype.prefixCls;
    /** @type {?} */
    AgreeItem.prototype.name;
    /** @type {?} */
    AgreeItem.prototype.value;
    /** @type {?} */
    AgreeItem.prototype.checked;
    /** @type {?} */
    AgreeItem.prototype.disabled;
    /** @type {?} */
    AgreeItem.prototype.onChange;
    /** @type {?} */
    AgreeItem.prototype.checkboxAgree;
}
