// ng-zorro-antd-mobile/esm2015/menu/sub-menu/sub-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, EventEmitter, } from '@angular/core';
export class SubMenuComponent {
    constructor() {
        this.prefixCls = 'am-sub-menu';
        this.onSel = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get subMenuPrefixCls() {
        return this._subMenuPrefixCls;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set subMenuPrefixCls(v) {
        this._subMenuPrefixCls = v;
    }
    /**
     * @return {?}
     */
    get radioPrefixCls() {
        return this._radioPrefixCls;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set radioPrefixCls(v) {
        this._radioPrefixCls = v;
    }
    /**
     * @return {?}
     */
    get subMenuData() {
        return this._subMenuData;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set subMenuData(v) {
        this._subMenuData = v;
    }
    /**
     * @return {?}
     */
    get showSelect() {
        return this._showSelect;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set showSelect(v) {
        this._showSelect = v;
    }
    /**
     * @return {?}
     */
    get selItem() {
        return this._selItem;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set selItem(v) {
        this._selItem = v;
    }
    /**
     * @return {?}
     */
    get multiSelect() {
        return this._multiSelect;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    set multiSelect(v) {
        this._multiSelect = v;
    }
    /**
     * @param {?} dataItem
     * @return {?}
     */
    onClick(dataItem) {
        this.onSel.emit(dataItem);
    }
    /**
     * @param {?} dataItem
     * @return {?}
     */
    selected(dataItem) {
        return this._showSelect && (this._selItem.length > 0 && this._selItem.indexOf(dataItem.value) !== -1);
    }
    /**
     * @param {?} dataItem
     * @return {?}
     */
    getClass(dataItem) {
        /** @type {?} */
        let name = this._radioPrefixCls + '-item ';
        name += this.selected(dataItem) ? this._subMenuPrefixCls + '-item-selected' : '';
        name += dataItem.disabled ? this._subMenuPrefixCls + '-item-disabled' : '';
        return name;
    }
}
SubMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'SubMenu, nzm-sub-menu',
                template: "<List class=\"{{subMenuPrefixCls}}\" style=\"padding: 0\">\n  <ListItem *ngFor=\"let dataItem of subMenuData; let i = index;\"\n            key=\"i\"\n            [className]=\"getClass(dataItem)\"\n            [extra]=\"extra\"\n  >\n    {{dataItem.label}}\n\n    <ng-template #extra>\n      <label Radio\n             *ngIf=\"!multiSelect\"\n             [checked]=\"selected(dataItem)\"\n             [disabled]=\"dataItem.disabled\"\n             (onChange)=\"onClick(dataItem)\"\n      >\n      </label>\n      <label Checkbox\n             *ngIf=\"multiSelect\"\n             [checked]=\"selected(dataItem)\"\n             [disabled]=\"dataItem.disabled\"\n             (onChange)=\"onClick(dataItem)\"\n      >\n      </label>\n    </ng-template>\n  </ListItem>\n</List>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
SubMenuComponent.ctorParameters = () => [];
SubMenuComponent.propDecorators = {
    onSel: [{ type: Output }],
    subMenuPrefixCls: [{ type: Input }],
    radioPrefixCls: [{ type: Input }],
    subMenuData: [{ type: Input }],
    showSelect: [{ type: Input }],
    selItem: [{ type: Input }],
    multiSelect: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    SubMenuComponent.prototype.prefixCls;
    /** @type {?} */
    SubMenuComponent.prototype._subMenuPrefixCls;
    /** @type {?} */
    SubMenuComponent.prototype._radioPrefixCls;
    /** @type {?} */
    SubMenuComponent.prototype._subMenuData;
    /** @type {?} */
    SubMenuComponent.prototype._showSelect;
    /** @type {?} */
    SubMenuComponent.prototype._selItem;
    /** @type {?} */
    SubMenuComponent.prototype._multiSelect;
    /** @type {?} */
    SubMenuComponent.prototype.onSel;
}
