// ng-zorro-antd-mobile/esm2015/calendar/week-panel/week-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation, Input } from '@angular/core';
export class WeekPanelComponent {
    constructor() {
        this.week = ['日', '一', '二', '三', '四', '五', '六'];
        this.weekPanel = true;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set locale(value) {
        this._locale = value;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.week = this._locale.week;
    }
}
WeekPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'CalendarWeekPanel, nzm-calendar-week-panel',
                template: "<div class=\"cell cell-grey\">{{week[0]}}</div>\n<div class=\"cell\">{{week[1]}}</div>\n<div class=\"cell\">{{week[2]}}</div>\n<div class=\"cell\">{{week[3]}}</div>\n<div class=\"cell\">{{week[4]}}</div>\n<div class=\"cell\">{{week[5]}}</div>\n<div class=\"cell cell-grey\">{{week[6]}}</div>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
WeekPanelComponent.ctorParameters = () => [];
WeekPanelComponent.propDecorators = {
    locale: [{ type: Input }],
    weekPanel: [{ type: HostBinding, args: ['class.week-panel',] }]
};
if (false) {
    /** @type {?} */
    WeekPanelComponent.prototype.week;
    /** @type {?} */
    WeekPanelComponent.prototype._locale;
    /** @type {?} */
    WeekPanelComponent.prototype.weekPanel;
}
