// ng-zorro-antd-mobile/esm5/menu/sub-menu/sub-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, EventEmitter, } from '@angular/core';
var SubMenuComponent = /** @class */ (function () {
    function SubMenuComponent() {
        this.prefixCls = 'am-sub-menu';
        this.onSel = new EventEmitter();
    }
    Object.defineProperty(SubMenuComponent.prototype, "subMenuPrefixCls", {
        get: /**
         * @return {?}
         */
        function () {
            return this._subMenuPrefixCls;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._subMenuPrefixCls = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubMenuComponent.prototype, "radioPrefixCls", {
        get: /**
         * @return {?}
         */
        function () {
            return this._radioPrefixCls;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._radioPrefixCls = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubMenuComponent.prototype, "subMenuData", {
        get: /**
         * @return {?}
         */
        function () {
            return this._subMenuData;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._subMenuData = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubMenuComponent.prototype, "showSelect", {
        get: /**
         * @return {?}
         */
        function () {
            return this._showSelect;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._showSelect = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubMenuComponent.prototype, "selItem", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selItem;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._selItem = v;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SubMenuComponent.prototype, "multiSelect", {
        get: /**
         * @return {?}
         */
        function () {
            return this._multiSelect;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._multiSelect = v;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} dataItem
     * @return {?}
     */
    SubMenuComponent.prototype.onClick = /**
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        this.onSel.emit(dataItem);
    };
    /**
     * @param {?} dataItem
     * @return {?}
     */
    SubMenuComponent.prototype.selected = /**
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        return this._showSelect && (this._selItem.length > 0 && this._selItem.indexOf(dataItem.value) !== -1);
    };
    /**
     * @param {?} dataItem
     * @return {?}
     */
    SubMenuComponent.prototype.getClass = /**
     * @param {?} dataItem
     * @return {?}
     */
    function (dataItem) {
        /** @type {?} */
        var name = this._radioPrefixCls + '-item ';
        name += this.selected(dataItem) ? this._subMenuPrefixCls + '-item-selected' : '';
        name += dataItem.disabled ? this._subMenuPrefixCls + '-item-disabled' : '';
        return name;
    };
    SubMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'SubMenu, nzm-sub-menu',
                    template: "<List class=\"{{subMenuPrefixCls}}\" style=\"padding: 0\">\n  <ListItem *ngFor=\"let dataItem of subMenuData; let i = index;\"\n            key=\"i\"\n            [className]=\"getClass(dataItem)\"\n            [extra]=\"extra\"\n  >\n    {{dataItem.label}}\n\n    <ng-template #extra>\n      <label Radio\n             *ngIf=\"!multiSelect\"\n             [checked]=\"selected(dataItem)\"\n             [disabled]=\"dataItem.disabled\"\n             (onChange)=\"onClick(dataItem)\"\n      >\n      </label>\n      <label Checkbox\n             *ngIf=\"multiSelect\"\n             [checked]=\"selected(dataItem)\"\n             [disabled]=\"dataItem.disabled\"\n             (onChange)=\"onClick(dataItem)\"\n      >\n      </label>\n    </ng-template>\n  </ListItem>\n</List>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    SubMenuComponent.ctorParameters = function () { return []; };
    SubMenuComponent.propDecorators = {
        onSel: [{ type: Output }],
        subMenuPrefixCls: [{ type: Input }],
        radioPrefixCls: [{ type: Input }],
        subMenuData: [{ type: Input }],
        showSelect: [{ type: Input }],
        selItem: [{ type: Input }],
        multiSelect: [{ type: Input }]
    };
    return SubMenuComponent;
}());
export { SubMenuComponent };
if (false) {
    /** @type {?} */
    SubMenuComponent.prototype.prefixCls;
    /** @type {?} */
    SubMenuComponent.prototype._subMenuPrefixCls;
    /** @type {?} */
    SubMenuComponent.prototype._radioPrefixCls;
    /** @type {?} */
    SubMenuComponent.prototype._subMenuData;
    /** @type {?} */
    SubMenuComponent.prototype._showSelect;
    /** @type {?} */
    SubMenuComponent.prototype._selItem;
    /** @type {?} */
    SubMenuComponent.prototype._multiSelect;
    /** @type {?} */
    SubMenuComponent.prototype.onSel;
}
