// ng-zorro-antd-mobile/esm5/carousel/carousel-slide/carousel-slide.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
var CarouselSlideComponent = /** @class */ (function () {
    function CarouselSlideComponent() {
        this.container = true;
        this.height = 'auto';
    }
    CarouselSlideComponent.decorators = [
        { type: Component, args: [{
                    selector: 'CarouselSlide, nzm-carousel-slide',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    CarouselSlideComponent.propDecorators = {
        container: [{ type: HostBinding, args: ['class.carousel-container',] }],
        width: [{ type: HostBinding, args: ['style.width.px',] }],
        height: [{ type: HostBinding, args: ['style.height',] }],
        left: [{ type: HostBinding, args: ['style.left.px',] }],
        top: [{ type: HostBinding, args: ['style.top.px',] }],
        margin: [{ type: HostBinding, args: ['style.margin',] }]
    };
    return CarouselSlideComponent;
}());
export { CarouselSlideComponent };
if (false) {
    /** @type {?} */
    CarouselSlideComponent.prototype.container;
    /** @type {?} */
    CarouselSlideComponent.prototype.width;
    /** @type {?} */
    CarouselSlideComponent.prototype.height;
    /** @type {?} */
    CarouselSlideComponent.prototype.left;
    /** @type {?} */
    CarouselSlideComponent.prototype.top;
    /** @type {?} */
    CarouselSlideComponent.prototype.margin;
}
