// ng-zorro-antd-mobile/esm5/list/list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
var List = /** @class */ (function () {
    function List() {
        this.defaultProps = {
            prefixCls: 'am-list'
        };
        this._renderHeader = '';
        this._renderFooter = '';
        this._className = '';
    }
    Object.defineProperty(List.prototype, "className", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._className = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(List.prototype, "renderHeader", {
        get: /**
         * @return {?}
         */
        function () {
            return this._renderHeader;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._renderHeader = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(List.prototype, "renderFooter", {
        get: /**
         * @return {?}
         */
        function () {
            return this._renderFooter;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._renderFooter = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(List.prototype, "hostClassName", {
        get: /**
         * @return {?}
         */
        function () {
            return 'am-list ' + this._className;
        },
        enumerable: true,
        configurable: true
    });
    List.decorators = [
        { type: Component, args: [{
                    selector: 'List, nzm-list',
                    template: "<div *ngIf=\"renderHeader\" class=\"{{defaultProps.prefixCls}}-header\">\n  {{renderHeader()}}\n</div>\n<div class=\"{{defaultProps.prefixCls}}-body\">\n  <ng-content></ng-content>\n</div>\n<div *ngIf=\"renderFooter\" class=\"{{defaultProps.prefixCls}}-footer\">\n  {{renderFooter()}}\n</div>\n\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    List.ctorParameters = function () { return []; };
    List.propDecorators = {
        className: [{ type: Input }],
        renderHeader: [{ type: Input }],
        renderFooter: [{ type: Input }],
        hostClassName: [{ type: HostBinding, args: ['class',] }]
    };
    return List;
}());
export { List };
if (false) {
    /** @type {?} */
    List.prototype.defaultProps;
    /** @type {?} */
    List.prototype._renderHeader;
    /** @type {?} */
    List.prototype._renderFooter;
    /** @type {?} */
    List.prototype._className;
}
