// ng-zorro-antd-mobile/esm5/core/directive/touch-feedback.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter, HostListener, ElementRef, Renderer2, InjectionToken, } from '@angular/core';
/** @type {?} */
export var INTERFACE_TOKEN = new InjectionToken('InterfaceToken');
var TouchFeedbackDirective = /** @class */ (function () {
    function TouchFeedbackDirective(_elementRef, _renderer) {
        this._elementRef = _elementRef;
        this._renderer = _renderer;
        this.activeStyle = true;
        this.clickStart = new EventEmitter();
        this.clickEnd = new EventEmitter();
    }
    /**
     * @param {?} className
     * @return {?}
     */
    TouchFeedbackDirective.prototype.addClass = /**
     * @param {?} className
     * @return {?}
     */
    function (className) {
        this._renderer.addClass(this._elementRef.nativeElement, className);
    };
    /**
     * @param {?} className
     * @return {?}
     */
    TouchFeedbackDirective.prototype.removeClass = /**
     * @param {?} className
     * @return {?}
     */
    function (className) {
        this._renderer.removeClass(this._elementRef.nativeElement, className);
    };
    /**
     * @return {?}
     */
    TouchFeedbackDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._className = this.className;
    };
    /**
     * @return {?}
     */
    TouchFeedbackDirective.prototype.touchStart = /**
     * @return {?}
     */
    function () {
        if (this.activeStyle) {
            this.addClass(this._className);
            this.clickStart.emit();
        }
    };
    /**
     * @return {?}
     */
    TouchFeedbackDirective.prototype.touchEnd = /**
     * @return {?}
     */
    function () {
        if (this.activeStyle) {
            this.removeClass(this._className);
            this.clickEnd.emit();
        }
    };
    TouchFeedbackDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[TouchFeedback]'
                },] }
    ];
    /** @nocollapse */
    TouchFeedbackDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    TouchFeedbackDirective.propDecorators = {
        className: [{ type: Input }],
        activeStyle: [{ type: Input }],
        clickStart: [{ type: Output }],
        clickEnd: [{ type: Output }],
        touchStart: [{ type: HostListener, args: ['touchstart', ['$event'],] }, { type: HostListener, args: ['mousedown', ['$event'],] }],
        touchEnd: [{ type: HostListener, args: ['touchend', ['$event'],] }, { type: HostListener, args: ['mouseup', ['$event'],] }]
    };
    return TouchFeedbackDirective;
}());
export { TouchFeedbackDirective };
if (false) {
    /** @type {?} */
    TouchFeedbackDirective.prototype._className;
    /** @type {?} */
    TouchFeedbackDirective.prototype.className;
    /** @type {?} */
    TouchFeedbackDirective.prototype.activeStyle;
    /** @type {?} */
    TouchFeedbackDirective.prototype.clickStart;
    /** @type {?} */
    TouchFeedbackDirective.prototype.clickEnd;
    /** @type {?} */
    TouchFeedbackDirective.prototype._elementRef;
    /** @type {?} */
    TouchFeedbackDirective.prototype._renderer;
}
