// ng-zorro-antd-mobile/esm5/textarea-item/textarea-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var TextareaItem = /** @class */ (function () {
    function TextareaItem() {
        this.prefixCls = 'am-textarea';
        this.isTitleString = true;
        this.maxLength = Infinity;
        this._prefixListCls = 'am-list';
        this._defaultValue = '';
        this._placeholder = '';
        this._editable = true;
        this._disabled = false;
        this._clear = false;
        this._rows = 1;
        this._error = false;
        this._labelNumber = 5;
        this._name = '';
        this._focus = false;
        this._autoFocus = false;
        this._isClear = false;
        this._isClickingClear = false;
        this.onChange = new EventEmitter();
        this.onBlur = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.onErrorClick = new EventEmitter();
    }
    Object.defineProperty(TextareaItem.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (typeof v === undefined || v === null) {
                this._value = '';
            }
            this._value = v;
            this.textRef.nativeElement.value = this._value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "defaultValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._defaultValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._defaultValue = value;
            this._value = this._defaultValue;
            this.textRef.nativeElement.value = this._value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "placeholder", {
        get: /**
         * @return {?}
         */
        function () {
            return this._placeholder;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._placeholder = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "editable", {
        get: /**
         * @return {?}
         */
        function () {
            return this._editable;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._editable = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "clear", {
        get: /**
         * @return {?}
         */
        function () {
            return this._clear;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._clear = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "rows", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rows;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._rows = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "error", {
        get: /**
         * @return {?}
         */
        function () {
            return this._error;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._error = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "labelNumber", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._labelNumber = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "count", {
        get: /**
         * @return {?}
         */
        function () {
            return this._count;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._count = value;
            this.setCls();
            this.setCharacterLength();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "prefixListCls", {
        get: /**
         * @return {?}
         */
        function () {
            return this._prefixListCls;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._prefixListCls = value;
            this.setCls();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "name", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._name = value;
            this.textRef.nativeElement.name = this._name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "autoHeight", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._autoHeight = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._title = value;
            this.isTitleString = true;
            if (typeof value !== 'string') {
                this.isTitleString = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "focus", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value && value.focus) {
                this.textRef.nativeElement.focus();
                this.inputFocus('');
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextareaItem.prototype, "autoFocus", {
        get: /**
         * @return {?}
         */
        function () {
            return this._autoFocus;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._autoFocus = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TextareaItem.prototype.setCls = /**
     * @return {?}
     */
    function () {
        var _a, _b, _c, _d;
        this.hasCount = this._count > 0 && this._rows > 1;
        this.wrapCls = (_a = {},
            _a[this._prefixListCls + "-item"] = true,
            _a[this.prefixCls + "-item"] = true,
            _a[this.prefixCls + "-disabled"] = this._disabled,
            _a[this.prefixCls + "-error"] = this._error,
            _a[this.prefixCls + "-focus"] = this._focus,
            _a[this.prefixCls + "-item-single-line"] = this._rows === 1 && !this._autoHeight,
            _a[this.prefixCls + "-has-count"] = this.hasCount,
            _a);
        this.labelCls = (_b = {},
            _b[this.prefixCls + "-label"] = true,
            _b[this.prefixCls + "-label-2"] = this._labelNumber === 2,
            _b[this.prefixCls + "-label-3"] = this._labelNumber === 3,
            _b[this.prefixCls + "-label-4"] = this._labelNumber === 4,
            _b[this.prefixCls + "-label-5"] = this._labelNumber === 5,
            _b[this.prefixCls + "-label-6"] = this._labelNumber === 6,
            _b[this.prefixCls + "-label-7"] = this._labelNumber === 7,
            _b);
        this.controlCls = (_c = {}, _c[this.prefixCls + "-control"] = true, _c);
        this.clearCls = (_d = {},
            _d[this.prefixCls + "-clear-active"] = this._isClickingClear,
            _d);
    };
    /**
     * @return {?}
     */
    TextareaItem.prototype.setCharacterLength = /**
     * @return {?}
     */
    function () {
        this.characterLength = this.countSymbols(this._value);
        if (this._count > 0) {
            this.maxLength = this._count - this.characterLength + (this._value ? this._value.length : 0);
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    TextareaItem.prototype.inputChange = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this._value = e;
        this.textRef.nativeElement.value = this._value;
        this.setCharacterLength();
        this.onChange.emit(this._value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    TextareaItem.prototype.inputFocus = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this._focus = true;
        this.setCls();
        if (value !== undefined) {
            this.onFocus.emit(value);
        }
    };
    /**
     * @param {?} value
     * @param {?} event
     * @return {?}
     */
    TextareaItem.prototype.inputBlur = /**
     * @param {?} value
     * @param {?} event
     * @return {?}
     */
    function (value, event) {
        var _this = this;
        setTimeout(function () {
            _this._focus = false;
            _this.setCls();
            _this.onBlur.emit(value);
            _this._isClear = false;
        }, 100);
    };
    /**
     * @return {?}
     */
    TextareaItem.prototype.clearInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._isClickingClear = true;
        this.setCls();
        setTimeout(function () {
            _this._value = '';
            _this.inputChange('');
            _this.inputFocus(_this._value);
            _this._isClickingClear = false;
            _this.setCls();
        }, 100);
    };
    /**
     * @param {?} e
     * @return {?}
     */
    TextareaItem.prototype.errorClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this.onErrorClick) {
            this.onErrorClick.emit(e);
        }
    };
    /**
     * @return {?}
     */
    TextareaItem.prototype.reAlignHeight = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var textareaDom = this.textRef.nativeElement;
        textareaDom.style.height = '';
        textareaDom.style.height = textareaDom.scrollHeight + "px";
    };
    /**
     * @param {?=} text
     * @return {?}
     */
    TextareaItem.prototype.countSymbols = /**
     * @param {?=} text
     * @return {?}
     */
    function (text) {
        if (text === void 0) { text = ''; }
        /** @type {?} */
        var regexAstralSymbols = /[\uD800-\uDBFF][\uDC00-\uDFFF]|\n/g;
        return text.replace(regexAstralSymbols, '_').length;
    };
    /**
     * @return {?}
     */
    TextareaItem.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setCls();
        this.setCharacterLength();
        this.textRef.nativeElement.value = this._value;
    };
    /**
     * @return {?}
     */
    TextareaItem.prototype.ngAfterContentChecked = /**
     * @return {?}
     */
    function () {
        if (this._autoHeight && this._focus) {
            this.reAlignHeight();
        }
    };
    TextareaItem.decorators = [
        { type: Component, args: [{
                    selector: 'TextareaItem , nzm-textarea-item',
                    template: "<div [ngClass]=\"wrapCls\">\n  <div *ngIf=\"title &&  isTitleString\" [ngClass]=\"labelCls\">{{title}}</div>\n  <div *ngIf=\"title &&  !isTitleString\" [ngClass]=\"labelCls\">\n    <ng-template [ngTemplateOutlet]=\"title\"></ng-template>\n  </div>\n  <div [ngClass]=\"controlCls\">\n    <textarea #text\n              [rows]=\"rows\"\n              [maxlength]=\"maxLength\"\n              [(ngModel)]=\"value\"\n              [defaultValue]=\"defaultValue\"\n              [placeholder]=\"placeholder\"\n              [disabled]=\"disabled\"\n              [readOnly]=\"!editable\"\n              [autofocus]=\"autoFocus\"\n              (ngModelChange)=\"inputChange($event)\"\n              (blur)=\"inputBlur(value , $event)\"\n              (focus)=\"inputFocus(value)\"\n    ></textarea>\n  </div>\n  <div *ngIf=\"clear && editable && !disabled && (value && value.length >0 )\"\n       class=\"{{prefixCls}}-clear\"\n       [ngClass]=\"clearCls\"\n       (click)=\"clearInput()\"\n  ></div>\n  <div *ngIf=\"error\" class=\"{{prefixCls}}-error-extra\" (click)=\"errorClick($event)\"></div>\n  <span *ngIf=\"hasCount\" class=\"{{prefixCls}}-count\">\n    <span>{{characterLength}}</span>/{{count}}\n  </span>\n</div>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return TextareaItem; }),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    TextareaItem.ctorParameters = function () { return []; };
    TextareaItem.propDecorators = {
        textRef: [{ type: ViewChild, args: ['text',] }],
        value: [{ type: Input }],
        defaultValue: [{ type: Input }],
        placeholder: [{ type: Input }],
        editable: [{ type: Input }],
        disabled: [{ type: Input }],
        clear: [{ type: Input }],
        rows: [{ type: Input }],
        error: [{ type: Input }],
        labelNumber: [{ type: Input }],
        count: [{ type: Input }],
        prefixListCls: [{ type: Input }],
        name: [{ type: Input }],
        autoHeight: [{ type: Input }],
        title: [{ type: Input }],
        focus: [{ type: Input }],
        autoFocus: [{ type: Input }],
        onChange: [{ type: Output }],
        onBlur: [{ type: Output }],
        onFocus: [{ type: Output }],
        onErrorClick: [{ type: Output }]
    };
    return TextareaItem;
}());
export { TextareaItem };
if (false) {
    /** @type {?} */
    TextareaItem.prototype.prefixCls;
    /** @type {?} */
    TextareaItem.prototype.wrapCls;
    /** @type {?} */
    TextareaItem.prototype.labelCls;
    /** @type {?} */
    TextareaItem.prototype.controlCls;
    /** @type {?} */
    TextareaItem.prototype.clearCls;
    /** @type {?} */
    TextareaItem.prototype.hasCount;
    /** @type {?} */
    TextareaItem.prototype.characterLength;
    /** @type {?} */
    TextareaItem.prototype.isTitleString;
    /** @type {?} */
    TextareaItem.prototype.maxLength;
    /** @type {?} */
    TextareaItem.prototype._prefixListCls;
    /** @type {?} */
    TextareaItem.prototype._value;
    /** @type {?} */
    TextareaItem.prototype._defaultValue;
    /** @type {?} */
    TextareaItem.prototype._placeholder;
    /** @type {?} */
    TextareaItem.prototype._editable;
    /** @type {?} */
    TextareaItem.prototype._disabled;
    /** @type {?} */
    TextareaItem.prototype._clear;
    /** @type {?} */
    TextareaItem.prototype._rows;
    /** @type {?} */
    TextareaItem.prototype._count;
    /** @type {?} */
    TextareaItem.prototype._autoHeight;
    /** @type {?} */
    TextareaItem.prototype._error;
    /** @type {?} */
    TextareaItem.prototype._labelNumber;
    /** @type {?} */
    TextareaItem.prototype._name;
    /** @type {?} */
    TextareaItem.prototype._title;
    /** @type {?} */
    TextareaItem.prototype._focus;
    /** @type {?} */
    TextareaItem.prototype._autoFocus;
    /** @type {?} */
    TextareaItem.prototype._isClear;
    /** @type {?} */
    TextareaItem.prototype._isClickingClear;
    /** @type {?} */
    TextareaItem.prototype.textRef;
    /** @type {?} */
    TextareaItem.prototype.onChange;
    /** @type {?} */
    TextareaItem.prototype.onBlur;
    /** @type {?} */
    TextareaItem.prototype.onFocus;
    /** @type {?} */
    TextareaItem.prototype.onErrorClick;
}
