// ng-zorro-antd-mobile/esm5/steps/step/step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input, ElementRef, ViewEncapsulation, HostBinding } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
var Step = /** @class */ (function () {
    function Step(_el) {
        this._el = _el;
        this.prefixCls = 'am-steps';
        this.stepItemCls = {};
        this.isIconString = true;
        this.clsStepItem = true;
    }
    Object.defineProperty(Step.prototype, "status", {
        get: /**
         * @return {?}
         */
        function () {
            return this._status;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value) {
                this._status = value;
                this.setClass();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Step.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._title = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Step.prototype, "description", {
        get: /**
         * @return {?}
         */
        function () {
            return this._description;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._description = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Step.prototype, "icon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._icon;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value) {
                if (typeof value !== 'string') {
                    this.isIconString = false;
                }
                this._icon = value;
                this.setClass();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Step.prototype.setClass = /**
     * @return {?}
     */
    function () {
        var _a, _b;
        this.iconCls = (_a = {},
            _a[this.prefixCls + "-icon"] = true,
            _a);
        this.stepItemCls = Object.assign(this.stepItemCls, (_b = {},
            _b[this.prefixCls + "-item-" + this.status] = true,
            _b[this.prefixCls + "-item-custom"] = this.icon,
            _b));
    };
    /**
     * @return {?}
     */
    Step.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    Step.decorators = [
        { type: Component, args: [{
                    selector: 'Step, nzm-step',
                    template: "<div [ngClass]=\"stepItemCls\">\n  <div class=\"{{prefixCls}}-item-tail\">\n    {{tailContent}}\n  </div>\n  <div class=\"{{prefixCls}}-item-icon\">\n    <span *ngIf=\"!isIconString\" [ngClass]=\"iconCls\">\n      <ng-template [ngTemplateOutlet]=\"icon\"></ng-template>\n    </span>\n    <span *ngIf=\"isIconString && (status=== 'error' || status==='finish' || status==='wait' )\" [ngClass]=\"iconCls\">\n      <Icon [type]=\"icon\" [size]=\"iconSize\">\n      </Icon>\n    </span>\n    <span *ngIf=\"isIconString && !(status=== 'error' || status==='finish' || status==='wait')\" [ngClass]=\"iconCls\">{{stepNumber}}\n    </span>\n  </div>\n  <div class=\"{{prefixCls}}-item-content\">\n    <div class=\"{{prefixCls}}-item-title\">\n      {{title}}\n    </div>\n    <div *ngIf=\"description\" class=\"{{prefixCls}}-item-description\">{{description}}</div>\n  </div>\n</div>\n",
                    encapsulation: ViewEncapsulation.None,
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return Step; }),
                            multi: true
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    Step.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    Step.propDecorators = {
        status: [{ type: Input }],
        title: [{ type: Input }],
        description: [{ type: Input }],
        icon: [{ type: Input }],
        clsStepItem: [{ type: HostBinding, args: ['class.am-steps-item',] }]
    };
    return Step;
}());
export { Step };
if (false) {
    /** @type {?} */
    Step.prototype.prefixCls;
    /** @type {?} */
    Step.prototype.stepItemCls;
    /** @type {?} */
    Step.prototype.iconCls;
    /** @type {?} */
    Step.prototype.tailContent;
    /** @type {?} */
    Step.prototype.stepNumber;
    /** @type {?} */
    Step.prototype.isIconString;
    /** @type {?} */
    Step.prototype._status;
    /** @type {?} */
    Step.prototype._title;
    /** @type {?} */
    Step.prototype._description;
    /** @type {?} */
    Step.prototype._icon;
    /** @type {?} */
    Step.prototype.clsStepItem;
    /** @type {?} */
    Step.prototype._el;
}
/** @enum {string} */
var StatusEnum = {
    WAIT: 'wait',
    PROCESS: 'process',
    FINISH: 'finish',
    ERROR: 'error',
};
export { StatusEnum };
/** @enum {string} */
var DirectionEnum = {
    VERTICAL: 'vertical',
    HORIZONTAL: 'horizontal',
};
export { DirectionEnum };
