// ng-zorro-antd-mobile/esm5/activity-indicator/activity-indicator.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from '@angular/core';
var ActivityIndicator = /** @class */ (function () {
    function ActivityIndicator() {
        this.prefixCls = 'am-activity-indicator';
        this.spinnerClass = {};
        this._size = 'small';
        this._toast = false;
        this._animating = true;
        this.clsActIndicator = true;
    }
    Object.defineProperty(ActivityIndicator.prototype, "animating", {
        get: /**
         * @return {?}
         */
        function () {
            return this._animating;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._animating = v;
            this.setClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ActivityIndicator.prototype, "size", {
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._size = v;
            this.setClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ActivityIndicator.prototype, "toast", {
        get: /**
         * @return {?}
         */
        function () {
            return this._toast;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._toast = v;
            this.setClass();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ActivityIndicator.prototype, "text", {
        get: /**
         * @return {?}
         */
        function () {
            return this._text;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this._text = v;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ActivityIndicator.prototype.setClass = /**
     * @return {?}
     */
    function () {
        var _a;
        if (this._animating) {
            this.clsActIndicator = true;
            this.clsActIndicatorToast = !!this._toast;
            this.clsActIndicatorLg = this._size === 'large';
            this.clsActIndicatorSm = this._size === 'small';
            this.spinnerClass = (_a = {},
                _a[this.prefixCls + "-spinner"] = true,
                _a[this.prefixCls + "-spinner-lg"] = !!this._toast || this._size === 'large',
                _a);
        }
        else {
            this.clsActIndicator = false;
            this.clsActIndicatorLg = false;
            this.clsActIndicatorSm = false;
            this.clsActIndicatorToast = false;
        }
    };
    /**
     * @return {?}
     */
    ActivityIndicator.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.setClass();
    };
    ActivityIndicator.decorators = [
        { type: Component, args: [{
                    selector: 'ActivityIndicator , nzm-ctivity-indicator',
                    template: "<div *ngIf=\"animating && toast && text && text.length>0\">\n  <div class=\"{{prefixCls}}-content\">\n    <span [ngClass]=\"spinnerClass\" aria-hidden=\"'true'\"></span>\n    <span class=\"{{prefixCls}}-toast\">{{text}}</span>\n  </div>\n</div>\n<div *ngIf=\"animating && toast && !text\">\n  <div class=\"{{prefixCls}}-content\">\n    <span [ngClass]=\"spinnerClass\" aria-label=\"'Loading'\"></span>\n  </div>\n</div>\n<div *ngIf=\"animating && !toast && text && text.length>0\">\n  <span [ngClass]=\"spinnerClass\" aria-hidden=\"true\"></span>\n  <span class=\"{{prefixCls}}-tip\">{{text}}</span>\n</div>\n<div *ngIf=\"animating &&! toast && !text\">\n  <span [ngClass]=\"spinnerClass\" aria-label=\"'loading'\"></span>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    ActivityIndicator.ctorParameters = function () { return []; };
    ActivityIndicator.propDecorators = {
        animating: [{ type: Input }],
        size: [{ type: Input }],
        toast: [{ type: Input }],
        text: [{ type: Input }],
        clsActIndicator: [{ type: HostBinding, args: ['class.am-activity-indicator',] }],
        clsActIndicatorToast: [{ type: HostBinding, args: ['class.am-activity-indicator-toast',] }],
        clsActIndicatorLg: [{ type: HostBinding, args: ['class.am-activity-indicator-lg',] }],
        clsActIndicatorSm: [{ type: HostBinding, args: ['class.am-activity-indicator-sm',] }]
    };
    return ActivityIndicator;
}());
export { ActivityIndicator };
if (false) {
    /** @type {?} */
    ActivityIndicator.prototype.prefixCls;
    /** @type {?} */
    ActivityIndicator.prototype.spinnerClass;
    /** @type {?} */
    ActivityIndicator.prototype._text;
    /** @type {?} */
    ActivityIndicator.prototype._size;
    /** @type {?} */
    ActivityIndicator.prototype._toast;
    /** @type {?} */
    ActivityIndicator.prototype._animating;
    /** @type {?} */
    ActivityIndicator.prototype.clsActIndicator;
    /** @type {?} */
    ActivityIndicator.prototype.clsActIndicatorToast;
    /** @type {?} */
    ActivityIndicator.prototype.clsActIndicatorLg;
    /** @type {?} */
    ActivityIndicator.prototype.clsActIndicatorSm;
}
