// ng-zorro-antd-mobile/esm5/calendar/header/header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, HostBinding, TemplateRef, EventEmitter } from '@angular/core';
var HeaderComponent = /** @class */ (function () {
    function HeaderComponent() {
        this.closeIcon_component = false;
        this._closeIcon = 'X';
        this.onCancel = new EventEmitter();
        this.onClear = new EventEmitter();
        this.header = true;
    }
    Object.defineProperty(HeaderComponent.prototype, "locale", {
        get: /**
         * @return {?}
         */
        function () {
            return this._locale;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._locale = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeaderComponent.prototype, "closeIcon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._closeIcon;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof TemplateRef) {
                this._closeIcon = value;
                this.closeIcon_component = true;
            }
            else {
                this._closeIcon = /** @type {?} */ (value);
                this.closeIcon_component = false;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeaderComponent.prototype, "showClear", {
        get: /**
         * @return {?}
         */
        function () {
            return this._showClear;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._showClear = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    HeaderComponent.prototype.triggerCancel = /**
     * @return {?}
     */
    function () {
        this.onCancel && this.onCancel.emit();
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.triggerClear = /**
     * @return {?}
     */
    function () {
        this.onClear && this.onClear.emit();
    };
    HeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'CalendarHeader, nzm-calendar-header',
                    template: "<span *ngIf=\"!closeIcon_component\" class=\"left\" (click)=\"triggerCancel()\" [innerHTML]=\"closeIcon\"></span>\n<span *ngIf=\"closeIcon_component\" class=\"left\" (click)=\"triggerCancel()\">\n  <ng-template [ngTemplateOutlet]=\"closeIcon\"></ng-template>\n</span>\n<span class=\"title\">{{title || locale.title}}</span>\n<span *ngIf=\"showClear\" class=\"right\" (click)=\"triggerClear()\">{{clearIcon || locale.clear}}</span>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    /** @nocollapse */
    HeaderComponent.ctorParameters = function () { return []; };
    HeaderComponent.propDecorators = {
        locale: [{ type: Input }],
        closeIcon: [{ type: Input }],
        showClear: [{ type: Input }],
        onCancel: [{ type: Output }],
        onClear: [{ type: Output }],
        header: [{ type: HostBinding, args: ['class.header',] }]
    };
    return HeaderComponent;
}());
export { HeaderComponent };
if (false) {
    /** @type {?} */
    HeaderComponent.prototype.title;
    /** @type {?} */
    HeaderComponent.prototype.closeIcon_component;
    /** @type {?} */
    HeaderComponent.prototype.clearIcon;
    /** @type {?} */
    HeaderComponent.prototype._locale;
    /** @type {?} */
    HeaderComponent.prototype._showClear;
    /** @type {?} */
    HeaderComponent.prototype._closeIcon;
    /** @type {?} */
    HeaderComponent.prototype.onCancel;
    /** @type {?} */
    HeaderComponent.prototype.onClear;
    /** @type {?} */
    HeaderComponent.prototype.header;
}
