// ng-zorro-antd-mobile/esm2015/list/list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, HostBinding } from '@angular/core';
export class List {
    constructor() {
        this.defaultProps = {
            prefixCls: 'am-list'
        };
        this._renderHeader = '';
        this._renderFooter = '';
        this._className = '';
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set className(value) {
        this._className = value;
    }
    /**
     * @return {?}
     */
    get renderHeader() {
        return this._renderHeader;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set renderHeader(value) {
        this._renderHeader = value;
    }
    /**
     * @return {?}
     */
    get renderFooter() {
        return this._renderFooter;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set renderFooter(value) {
        this._renderFooter = value;
    }
    /**
     * @return {?}
     */
    get hostClassName() {
        return 'am-list ' + this._className;
    }
}
List.decorators = [
    { type: Component, args: [{
                selector: 'List, nzm-list',
                template: "<div *ngIf=\"renderHeader\" class=\"{{defaultProps.prefixCls}}-header\">\n  {{renderHeader()}}\n</div>\n<div class=\"{{defaultProps.prefixCls}}-body\">\n  <ng-content></ng-content>\n</div>\n<div *ngIf=\"renderFooter\" class=\"{{defaultProps.prefixCls}}-footer\">\n  {{renderFooter()}}\n</div>\n\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
List.ctorParameters = () => [];
List.propDecorators = {
    className: [{ type: Input }],
    renderHeader: [{ type: Input }],
    renderFooter: [{ type: Input }],
    hostClassName: [{ type: HostBinding, args: ['class',] }]
};
if (false) {
    /** @type {?} */
    List.prototype.defaultProps;
    /** @type {?} */
    List.prototype._renderHeader;
    /** @type {?} */
    List.prototype._renderFooter;
    /** @type {?} */
    List.prototype._className;
}
