// ng-zorro-antd-mobile/esm2015/wing-blank/wing-blank.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, ViewEncapsulation } from '@angular/core';
/** @typedef {?} */
var sizeType;
export { sizeType };
export class WingBlank {
    constructor() {
        this.prefixCls = 'am-wingblank';
        this.size = 'lg';
        this.amWingBlank = true;
    }
    /**
     * @return {?}
     */
    get amWingBlnkSm() {
        return this.size === 'sm';
    }
    /**
     * @return {?}
     */
    get amWingBlnkMd() {
        return this.size === 'md';
    }
    /**
     * @return {?}
     */
    get amWingBlnkLg() {
        return this.size === 'lg';
    }
}
WingBlank.decorators = [
    { type: Component, args: [{
                selector: 'WingBlank, nzm-wingblank',
                template: "<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
WingBlank.ctorParameters = () => [];
WingBlank.propDecorators = {
    size: [{ type: Input }],
    amWingBlank: [{ type: HostBinding, args: ['class.am-wingblank',] }],
    amWingBlnkSm: [{ type: HostBinding, args: ['class.am-wingblank-sm',] }],
    amWingBlnkMd: [{ type: HostBinding, args: ['class.am-wingblank-md',] }],
    amWingBlnkLg: [{ type: HostBinding, args: ['class.am-wingblank-lg',] }]
};
if (false) {
    /** @type {?} */
    WingBlank.prototype.prefixCls;
    /** @type {?} */
    WingBlank.prototype.size;
    /** @type {?} */
    WingBlank.prototype.amWingBlank;
}
