// ng-zorro-antd-mobile/esm2015/calendar/header/header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ViewEncapsulation, Input, Output, HostBinding, TemplateRef, EventEmitter } from '@angular/core';
export class HeaderComponent {
    constructor() {
        this.closeIcon_component = false;
        this._closeIcon = 'X';
        this.onCancel = new EventEmitter();
        this.onClear = new EventEmitter();
        this.header = true;
    }
    /**
     * @return {?}
     */
    get locale() {
        return this._locale;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set locale(value) {
        this._locale = value;
    }
    /**
     * @return {?}
     */
    get closeIcon() {
        return this._closeIcon;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set closeIcon(value) {
        if (value instanceof TemplateRef) {
            this._closeIcon = value;
            this.closeIcon_component = true;
        }
        else {
            this._closeIcon = /** @type {?} */ (value);
            this.closeIcon_component = false;
        }
    }
    /**
     * @return {?}
     */
    get showClear() {
        return this._showClear;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set showClear(value) {
        this._showClear = value;
    }
    /**
     * @return {?}
     */
    triggerCancel() {
        this.onCancel && this.onCancel.emit();
    }
    /**
     * @return {?}
     */
    triggerClear() {
        this.onClear && this.onClear.emit();
    }
}
HeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'CalendarHeader, nzm-calendar-header',
                template: "<span *ngIf=\"!closeIcon_component\" class=\"left\" (click)=\"triggerCancel()\" [innerHTML]=\"closeIcon\"></span>\n<span *ngIf=\"closeIcon_component\" class=\"left\" (click)=\"triggerCancel()\">\n  <ng-template [ngTemplateOutlet]=\"closeIcon\"></ng-template>\n</span>\n<span class=\"title\">{{title || locale.title}}</span>\n<span *ngIf=\"showClear\" class=\"right\" (click)=\"triggerClear()\">{{clearIcon || locale.clear}}</span>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
HeaderComponent.ctorParameters = () => [];
HeaderComponent.propDecorators = {
    locale: [{ type: Input }],
    closeIcon: [{ type: Input }],
    showClear: [{ type: Input }],
    onCancel: [{ type: Output }],
    onClear: [{ type: Output }],
    header: [{ type: HostBinding, args: ['class.header',] }]
};
if (false) {
    /** @type {?} */
    HeaderComponent.prototype.title;
    /** @type {?} */
    HeaderComponent.prototype.closeIcon_component;
    /** @type {?} */
    HeaderComponent.prototype.clearIcon;
    /** @type {?} */
    HeaderComponent.prototype._locale;
    /** @type {?} */
    HeaderComponent.prototype._showClear;
    /** @type {?} */
    HeaderComponent.prototype._closeIcon;
    /** @type {?} */
    HeaderComponent.prototype.onCancel;
    /** @type {?} */
    HeaderComponent.prototype.onClear;
    /** @type {?} */
    HeaderComponent.prototype.header;
}
