// ng-zorro-antd/esm5/timeline/nz-timeline-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Input, Renderer2, TemplateRef, ViewChild } from '@angular/core';
var NzTimelineItemComponent = /** @class */ (function () {
    function NzTimelineItemComponent(renderer) {
        this.renderer = renderer;
        this._color = 'blue';
        this._isLast = false;
    }
    Object.defineProperty(NzTimelineItemComponent.prototype, "isLast", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isLast;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isLast = value;
            if (this.isLast) {
                this.renderer.addClass(this.liTemplate.nativeElement, 'ant-timeline-item-last');
            }
            else {
                this.renderer.removeClass(this.liTemplate.nativeElement, 'ant-timeline-item-last');
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzTimelineItemComponent.prototype, "nzDot", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dot;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.isDotString = !(value instanceof TemplateRef);
            this._dot = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NzTimelineItemComponent.prototype, "nzColor", {
        get: /**
         * @return {?}
         */
        function () {
            return this._color;
        },
        set: /**
         * @param {?} color
         * @return {?}
         */
        function (color) {
            this._color = color;
            this.updateClassMap();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NzTimelineItemComponent.prototype.updateClassMap = /**
     * @return {?}
     */
    function () {
        var _a;
        /** @type {?} */
        var defaultColors = ['blue', 'red', 'green'];
        /** @type {?} */
        var circle = this.liTemplate.nativeElement.querySelector('.ant-timeline-item-head');
        if (defaultColors.indexOf(this._color) === -1) {
            this.renderer.setStyle(circle, 'border-color', this._color);
        }
        else {
            this.renderer.removeStyle(circle, 'border-color');
        }
        this.classMap = (_a = {},
            _a['ant-timeline-item-head-green'] = this.nzColor === 'green',
            _a['ant-timeline-item-head-red'] = this.nzColor === 'red',
            _a['ant-timeline-item-head-blue'] = this.nzColor === 'blue',
            _a);
    };
    /**
     * @return {?}
     */
    NzTimelineItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.updateClassMap();
    };
    NzTimelineItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'nz-timeline-item',
                    preserveWhitespaces: false,
                    template: "<li class=\"ant-timeline-item\" #liTemplate>\n  <div class=\"ant-timeline-item-tail\"></div>\n  <div\n    class=\"ant-timeline-item-head\"\n    [class.ant-timeline-item-head-custom]=\"nzDot\"\n    [ngClass]=\"classMap\">\n    <ng-container *ngIf=\"isDotString; else dotTemplate\">{{ nzDot }}</ng-container>\n    <ng-template #dotTemplate>\n      <ng-template [ngTemplateOutlet]=\"nzDot\"></ng-template>\n    </ng-template>\n  </div>\n  <div class=\"ant-timeline-item-content\">\n    <ng-content></ng-content>\n  </div>\n</li>"
                }] }
    ];
    /** @nocollapse */
    NzTimelineItemComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    NzTimelineItemComponent.propDecorators = {
        liTemplate: [{ type: ViewChild, args: ['liTemplate',] }],
        nzDot: [{ type: Input }],
        nzColor: [{ type: Input }]
    };
    return NzTimelineItemComponent;
}());
export { NzTimelineItemComponent };
function NzTimelineItemComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    NzTimelineItemComponent.prototype._dot;
    /** @type {?} */
    NzTimelineItemComponent.prototype._color;
    /** @type {?} */
    NzTimelineItemComponent.prototype._isLast;
    /** @type {?} */
    NzTimelineItemComponent.prototype.liTemplate;
    /** @type {?} */
    NzTimelineItemComponent.prototype.isDotString;
    /** @type {?} */
    NzTimelineItemComponent.prototype.classMap;
    /** @type {?} */
    NzTimelineItemComponent.prototype.renderer;
}
