// ng-zorro-antd-mobile/esm2015/core/directive/touch-feedback.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, HostListener, ViewEncapsulation, ElementRef, Renderer2 } from '@angular/core';
export class TouchFeedBackComponent {
    /**
     * @param {?} _ele
     * @param {?} _render
     */
    constructor(_ele, _render) {
        this._ele = _ele;
        this._render = _render;
        this.activeClassName = '';
        this.activeStyle = true;
    }
    /**
     * @return {?}
     */
    touchStart() {
        if (this.activeStyle) {
            this._render.addClass(this._ele.nativeElement, this.activeClassName);
        }
    }
    /**
     * @return {?}
     */
    touchEnd() {
        if (this.activeStyle) {
            this._render.removeClass(this._ele.nativeElement, this.activeClassName);
        }
    }
}
TouchFeedBackComponent.decorators = [
    { type: Component, args: [{
                selector: 'TouchFeedBack',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None
            }] }
];
/** @nocollapse */
TouchFeedBackComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TouchFeedBackComponent.propDecorators = {
    activeClassName: [{ type: Input }],
    activeStyle: [{ type: Input }],
    touchStart: [{ type: HostListener, args: ['touchstart', ['$event'],] }, { type: HostListener, args: ['mousedown', ['$event'],] }],
    touchEnd: [{ type: HostListener, args: ['touchend', ['$event'],] }, { type: HostListener, args: ['mouseup', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    TouchFeedBackComponent.prototype.activeClassName;
    /** @type {?} */
    TouchFeedBackComponent.prototype.activeStyle;
    /** @type {?} */
    TouchFeedBackComponent.prototype._ele;
    /** @type {?} */
    TouchFeedBackComponent.prototype._render;
}
