// ng-zorro-antd-mobile/esm5/action-sheet/action-sheet-subject.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/** @enum {number} */
var actionSheetEvent = {
    onShow: 0,
    onShown: 1,
    onHide: 2,
    onHidden: 3,
    onOk: 4,
    onCancel: 5,
    onDestroy: 6,
};
var ActionSheetSubject = /** @class */ (function (_super) {
    tslib_1.__extends(ActionSheetSubject, _super);
    function ActionSheetSubject() {
        var _this = _super.call(this) || this;
        _this._eventsQueue = {};
        _this.subscribe(function (value) {
            /** @type {?} */
            var eventQueue = _this._eventsQueue[value] || [];
            eventQueue.forEach(function (cb) {
                if (cb !== null && cb !== undefined) {
                    cb();
                }
            });
        });
        return _this;
    }
    /**
     * @param {?=} type
     * @return {?}
     */
    ActionSheetSubject.prototype.destroy = /**
     * @param {?=} type
     * @return {?}
     */
    function (type) {
        if (type === void 0) { type = 'onCancel'; }
        if (!this.isStopped && !this.closed) {
            this.next(type);
        }
    };
    /**
     * @param {?} eventType
     * @param {?} cb
     * @return {?}
     */
    ActionSheetSubject.prototype.on = /**
     * @param {?} eventType
     * @param {?} cb
     * @return {?}
     */
    function (eventType, cb) {
        if (this._eventsQueue[eventType]) {
            this._eventsQueue[eventType].push(cb);
        }
        else {
            this._eventsQueue[eventType] = [cb];
        }
    };
    ActionSheetSubject.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ActionSheetSubject.ctorParameters = function () { return []; };
    return ActionSheetSubject;
}(Subject));
export { ActionSheetSubject };
if (false) {
    /** @type {?} */
    ActionSheetSubject.prototype.modalId;
    /** @type {?} */
    ActionSheetSubject.prototype._eventsQueue;
}
